{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import sys, os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.special import comb\n",
    "from scipy import stats\n",
    "import scipy.cluster.hierarchy as hac\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('classic')\n",
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from collections import Counter\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "sns.set(rc={'figure.figsize':(15,8)})\n",
    "sns.set_context('poster')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "#import CCPA_lib as cp\n",
    "import sorf_lib as sf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context('poster')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "![](http://oregonstate.edu/instruct/bb450/fall14/stryer7/2/table_02_02.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.uniprot.org/uniprot/Q7V735\n",
    "\n",
    "http://tigrfams.jcvi.org/cgi-bin/HmmReportPage.cgi?acc=TIGR03798\n",
    "\n",
    "https://www.ebi.ac.uk/training/online/course/interpro-functional-and-structural-analysis-protei/sequence-searching/searching-interpro-batc\n",
    "\n",
    "http://www.ebi.ac.uk/interpro/sequencesearch/iprscan5-S20190707-131508-0462-76111813-p1m\n",
    "\n",
    "https://www.ebi.ac.uk/Tools/services/rest/iprscan5/result/iprscan5-S20190707-131508-0462-76111813-p1m/json"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "6 frame translation\n",
    "classify ORFs into:\n",
    "* known gene\n",
    "* inside gene (part of gene/antisense)\n",
    "* extending gene\n",
    "* outside of gene (independent/operon?)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_df = sf.load_genome('MIT9313')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>function</th>\n",
       "      <th>aliases</th>\n",
       "      <th>figfam</th>\n",
       "      <th>evidence_codes</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "      <th>genome</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>PMIT9313_0001</td>\n",
       "      <td>fig|1218.91.peg.1</td>\n",
       "      <td>peg</td>\n",
       "      <td>MIT9313_174_1340</td>\n",
       "      <td>174</td>\n",
       "      <td>1340</td>\n",
       "      <td>+</td>\n",
       "      <td>DNA polymerase III beta subunit (EC 2.7.7.7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG00066425</td>\n",
       "      <td>isu;DNA-replication isu;DNA_replication_cluster_1</td>\n",
       "      <td>atgaaactggtttgttcacaggctgaactcaatacagcactacagc...</td>\n",
       "      <td>MKLVCSQAELNTALQLVSRAVASRPTHPVLANVLLTADAGTDRLSL...</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>174</td>\n",
       "      <td>1340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>PMIT9313_0002</td>\n",
       "      <td>fig|1218.91.peg.2</td>\n",
       "      <td>peg</td>\n",
       "      <td>MIT9313_1368_2120</td>\n",
       "      <td>1368</td>\n",
       "      <td>2120</td>\n",
       "      <td>+</td>\n",
       "      <td>RNA metabolism-related protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ttgagtgatttgttgcatcatcgtgtacgttgtgatcaggggttag...</td>\n",
       "      <td>MSDLLHHRVRCDQGLDHGPGVLPWMHPPVHRLLGWVSRPSALRVSR...</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>1368</td>\n",
       "      <td>2120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>PMIT9313_0003</td>\n",
       "      <td>fig|1218.91.peg.3</td>\n",
       "      <td>peg</td>\n",
       "      <td>MIT9313_2160_4562</td>\n",
       "      <td>2160</td>\n",
       "      <td>4562</td>\n",
       "      <td>+</td>\n",
       "      <td>Phosphoribosylformylglycinamidine synthase, sy...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG01303876</td>\n",
       "      <td>icw(1);De_Novo_Purine_Biosynthesis</td>\n",
       "      <td>gtgtgtcagggccccgatttgagagttgactatgacgtggctgcgg...</td>\n",
       "      <td>MCQGPDLRVDYDVAAALRHEGLKPHDYDEICRRLQRAPNRVELGMF...</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>2160</td>\n",
       "      <td>4562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>PMIT9313_0004</td>\n",
       "      <td>fig|1218.91.peg.4</td>\n",
       "      <td>peg</td>\n",
       "      <td>MIT9313_4622_6079</td>\n",
       "      <td>4622</td>\n",
       "      <td>6079</td>\n",
       "      <td>+</td>\n",
       "      <td>Amidophosphoribosyltransferase (EC 2.4.2.14)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG00000179</td>\n",
       "      <td>isu;De_Novo_Purine_Biosynthesis</td>\n",
       "      <td>atgtgtggcattgttggcattgtttctactgcgctggtcaatcagc...</td>\n",
       "      <td>MCGIVGIVSTALVNQQIYDSLLLLQHRGQDSTGIATMDGSVFHLHK...</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>4622</td>\n",
       "      <td>6079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>PMIT9313_0005</td>\n",
       "      <td>fig|1218.91.peg.5</td>\n",
       "      <td>peg</td>\n",
       "      <td>MIT9313_8604_6115</td>\n",
       "      <td>8604</td>\n",
       "      <td>6115</td>\n",
       "      <td>-</td>\n",
       "      <td>DNA gyrase subunit A (EC 5.99.1.3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG00000080</td>\n",
       "      <td>idu(1);DNA_gyrase_subunits idu(1);DNA_topoisom...</td>\n",
       "      <td>atggccgaggagcgccttcaaccaatcgcattgcatcaagagatgc...</td>\n",
       "      <td>MAEERLQPIALHQEMQRSYLEYAMSVIVGRALPDARDGLKPVQRRI...</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>6115</td>\n",
       "      <td>8604</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  contig_id        gene_id         feature_id type           location  start  \\\n",
       "0   MIT9313  PMIT9313_0001  fig|1218.91.peg.1  peg   MIT9313_174_1340    174   \n",
       "1   MIT9313  PMIT9313_0002  fig|1218.91.peg.2  peg  MIT9313_1368_2120   1368   \n",
       "2   MIT9313  PMIT9313_0003  fig|1218.91.peg.3  peg  MIT9313_2160_4562   2160   \n",
       "3   MIT9313  PMIT9313_0004  fig|1218.91.peg.4  peg  MIT9313_4622_6079   4622   \n",
       "4   MIT9313  PMIT9313_0005  fig|1218.91.peg.5  peg  MIT9313_8604_6115   8604   \n",
       "\n",
       "   stop strand                                           function  aliases  \\\n",
       "0  1340      +       DNA polymerase III beta subunit (EC 2.7.7.7)      NaN   \n",
       "1  2120      +                     RNA metabolism-related protein      NaN   \n",
       "2  4562      +  Phosphoribosylformylglycinamidine synthase, sy...      NaN   \n",
       "3  6079      +       Amidophosphoribosyltransferase (EC 2.4.2.14)      NaN   \n",
       "4  6115      -                 DNA gyrase subunit A (EC 5.99.1.3)      NaN   \n",
       "\n",
       "        figfam                                     evidence_codes  \\\n",
       "0  FIG00066425  isu;DNA-replication isu;DNA_replication_cluster_1   \n",
       "1          NaN                                                NaN   \n",
       "2  FIG01303876                 icw(1);De_Novo_Purine_Biosynthesis   \n",
       "3  FIG00000179                    isu;De_Novo_Purine_Biosynthesis   \n",
       "4  FIG00000080  idu(1);DNA_gyrase_subunits idu(1);DNA_topoisom...   \n",
       "\n",
       "                                 nucleotide_sequence  \\\n",
       "0  atgaaactggtttgttcacaggctgaactcaatacagcactacagc...   \n",
       "1  ttgagtgatttgttgcatcatcgtgtacgttgtgatcaggggttag...   \n",
       "2  gtgtgtcagggccccgatttgagagttgactatgacgtggctgcgg...   \n",
       "3  atgtgtggcattgttggcattgtttctactgcgctggtcaatcagc...   \n",
       "4  atggccgaggagcgccttcaaccaatcgcattgcatcaagagatgc...   \n",
       "\n",
       "                                         aa_sequence   genome  left  right  \n",
       "0  MKLVCSQAELNTALQLVSRAVASRPTHPVLANVLLTADAGTDRLSL...  MIT9313   174   1340  \n",
       "1  MSDLLHHRVRCDQGLDHGPGVLPWMHPPVHRLLGWVSRPSALRVSR...  MIT9313  1368   2120  \n",
       "2  MCQGPDLRVDYDVAAALRHEGLKPHDYDEICRRLQRAPNRVELGMF...  MIT9313  2160   4562  \n",
       "3  MCGIVGIVSTALVNQQIYDSLLLLQHRGQDSTGIATMDGSVFHLHK...  MIT9313  4622   6079  \n",
       "4  MAEERLQPIALHQEMQRSYLEYAMSVIVGRALPDARDGLKPVQRRI...  MIT9313  6115   8604  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get all sequence records for the specified genbank file\n",
    "recs = [rec for rec in SeqIO.parse(os.path.join('data', \"MIT9313.gb\"), \"genbank\")]\n",
    "\n",
    "# print the number of sequence records that were extracted\n",
    "len(recs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(FeatureLocation(ExactPosition(173), ExactPosition(1340), strand=1),\n",
       " OrderedDict([('locus_tag', ['AKG35_RS00005']),\n",
       "              ('old_locus_tag', ['PMT0001', 'RG24_RS00005']),\n",
       "              ('EC_number', ['2.7.7.7']),\n",
       "              ('inference',\n",
       "               ['COORDINATES: similar to AA sequence:RefSeq:WP_011129380.1']),\n",
       "              ('note',\n",
       "               ['Derived by automated computational analysis using gene prediction method: Protein Homology.']),\n",
       "              ('codon_start', ['1']),\n",
       "              ('transl_table', ['11']),\n",
       "              ('product', ['DNA polymerase III subunit beta']),\n",
       "              ('protein_id', ['WP_011129380.1']),\n",
       "              ('db_xref', ['GI:499441916']),\n",
       "              ('translation',\n",
       "               ['MKLVCSQAELNTALQLVSRAVASRPTHPVLANVLLTADAGTDRLSLTGFDLSLGIQTALTASVESSGAITLPARLLGEIVARLSSDSPITLATDEAGEQVELKSSSGSYQMRGMSADDFPELPLVESGKAVKVNARALLTALRGTLFASSSDEAKQLLTGVHLHFDGKAMEAAATDGHRLAVLSLADALAVETTVSSEMDDEGENFSVTLPSRSLREVERLIAGWRGDDQVSLFCDKGQVVFLAADQVVTSRTLDGTYPNYRQLIPDGFARSFDVDRRAFIAALERIAVLADQHNNVVKVSGDNTSELLQISADAQDVGSGSESLSAEFTGDAVQIAFNVRYVLDGLKVMDSDRIVLRCNAPTTPAIISPKDDDIGFTYLVMPVQIRS'])]),\n",
       " 1)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recs[0].features[1].location, recs[0].features[2].qualifiers, recs[0].features[2].strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(FeatureLocation(ExactPosition(173), ExactPosition(1340), strand=1),\n",
       " OrderedDict([('locus_tag', ['AKG35_RS00005']),\n",
       "              ('old_locus_tag', ['PMT0001', 'RG24_RS00005'])]),\n",
       " 1)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recs[0].features[1].location, recs[0].features[1].qualifiers, recs[0].features[2].strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(FeatureLocation(ExactPosition(1343), ExactPosition(2120), strand=1),\n",
       " OrderedDict([('locus_tag', ['AKG35_RS00010']),\n",
       "              ('old_locus_tag', ['PMT0002', 'RG24_RS00010']),\n",
       "              ('inference',\n",
       "               ['COORDINATES: similar to AA sequence:RefSeq:WP_011824692.1']),\n",
       "              ('note',\n",
       "               ['Derived by automated computational analysis using gene prediction method: Protein Homology.']),\n",
       "              ('codon_start', ['1']),\n",
       "              ('transl_table', ['11']),\n",
       "              ('product', ['hypothetical protein']),\n",
       "              ('protein_id', ['WP_041384224.1']),\n",
       "              ('db_xref', ['GI:752737102']),\n",
       "              ('translation',\n",
       "               ['MNLPDQILLSDLLHHRVRCDQGLDHGPGVLPWMHPPVHRLLGWVSRPSALRVSRDVWRLDQSRGINEQEVFVKGQPAVSDQITLERLPTLLDADLLDRDGHRLGLVADFVFIPTTGKILHYLVSRSDPRLPGSSRWRLTTDRIVDQQPGIVATALRGLDDLPLVHSSVRQGLIKRSRHWRDQLHKMGDRASDRLEGWLEDPPWDQPVERSWKSSDMAEMDPLESWDDEQTDNFSQSTEELRHRGRPDGRGDGQDDPWV'])]),\n",
       " 1)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recs[0].features[4].location, recs[0].features[4].qualifiers, recs[0].features[4].strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1340"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int(recs[0].features[1].location.end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['contig_id', 'gene_id', 'feature_id', 'type', 'location', 'start',\n",
       "       'stop', 'strand', 'function', 'aliases', 'figfam', 'evidence_codes',\n",
       "       'nucleotide_sequence', 'aa_sequence', 'genome', 'left', 'right'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CDS', 'gene', 'ncRNA', 'rRNA', 'regulatory', 'source', 'tRNA', 'tmRNA'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{i.type for i in recs[0].features}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = recs[0].features[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('locus_tag', ['AKG35_RS00005']),\n",
       "             ('old_locus_tag', ['PMT0001', 'RG24_RS00005']),\n",
       "             ('EC_number', ['2.7.7.7']),\n",
       "             ('inference',\n",
       "              ['COORDINATES: similar to AA sequence:RefSeq:WP_011129380.1']),\n",
       "             ('note',\n",
       "              ['Derived by automated computational analysis using gene prediction method: Protein Homology.']),\n",
       "             ('codon_start', ['1']),\n",
       "             ('transl_table', ['11']),\n",
       "             ('product', ['DNA polymerase III subunit beta']),\n",
       "             ('protein_id', ['WP_011129380.1']),\n",
       "             ('db_xref', ['GI:499441916']),\n",
       "             ('translation',\n",
       "              ['MKLVCSQAELNTALQLVSRAVASRPTHPVLANVLLTADAGTDRLSLTGFDLSLGIQTALTASVESSGAITLPARLLGEIVARLSSDSPITLATDEAGEQVELKSSSGSYQMRGMSADDFPELPLVESGKAVKVNARALLTALRGTLFASSSDEAKQLLTGVHLHFDGKAMEAAATDGHRLAVLSLADALAVETTVSSEMDDEGENFSVTLPSRSLREVERLIAGWRGDDQVSLFCDKGQVVFLAADQVVTSRTLDGTYPNYRQLIPDGFARSFDVDRRAFIAALERIAVLADQHNNVVKVSGDNTSELLQISADAQDVGSGSESLSAEFTGDAVQIAFNVRYVLDGLKVMDSDRIVLRCNAPTTPAIISPKDDDIGFTYLVMPVQIRS'])])"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i.qualifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "    merge_cols = ['contig_id', 'gene_id', 'location', 'type', \n",
    "                  'start', 'stop', 'strand',\n",
    "                  'function', 'figfam',\n",
    "                  'nucleotide_sequence', 'aa_sequence', 'genome', 'left', 'right']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def gb_to_record(i, genome='MIT9313'):\n",
    "    q = i.qualifiers\n",
    "    g = lambda x : ','.join(q.get(x, []))\n",
    "    pmtid = [i for i in q.get('old_locus_tag',[]) if i.startswith('PMT')]\n",
    "    record = {\n",
    "        'contig_id' : genome, \n",
    "        'gene_id' : g('locus_tag'), \n",
    "        'pmt_id': ','.join(pmtid), \n",
    "        'type' : i.type, \n",
    "        'location' : g('locus_tag'),  \n",
    "        'strand' : '+' if i.location.strand == 1 else '-',\n",
    "        'start' : int(i.location.start),\n",
    "        'stop' : int(i.location.end),\n",
    "        'left' : int(i.location.start),\n",
    "        'right' : int(i.location.end),\n",
    "        'function' : g('product'), \n",
    "        'genome' : genome, \n",
    "        'old_locus_tag' : g('old_locus_tag'), \n",
    "        'product' : g('product'), \n",
    "        'db_xref' : g('db_xref'), \n",
    "        'protein_id' : g('protein_id'), \n",
    "        'figfam' : '',\n",
    "        'nucleotide_sequence' : '', \n",
    "        'aa_sequence' : g('translation'), \n",
    "    }\n",
    "    return record\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "types_to_collect = ['CDS', 'ncRNA', 'rRNA', 'regulatory', 'tRNA', 'tmRNA']\n",
    "records = [gb_to_record(i) for i in recs[0].features if i.type in types_to_collect]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = pd.DataFrame(records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf.to_feather(os.path.join('data', f'{genome}.ncbi.gb.feather.gz'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>pmt_id</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>function</th>\n",
       "      <th>genome</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>figfam</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS00225</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS00225</td>\n",
       "      <td>-</td>\n",
       "      <td>42281</td>\n",
       "      <td>42353</td>\n",
       "      <td>42281</td>\n",
       "      <td>42353</td>\n",
       "      <td>tRNA-Gly</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS00225,RNA_44</td>\n",
       "      <td>tRNA-Gly</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS00430</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS00430</td>\n",
       "      <td>+</td>\n",
       "      <td>87819</td>\n",
       "      <td>87891</td>\n",
       "      <td>87819</td>\n",
       "      <td>87891</td>\n",
       "      <td>tRNA-Gly</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS00430,RNA_1</td>\n",
       "      <td>tRNA-Gly</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regulatory</td>\n",
       "      <td></td>\n",
       "      <td>+</td>\n",
       "      <td>178355</td>\n",
       "      <td>178459</td>\n",
       "      <td>178355</td>\n",
       "      <td>178459</td>\n",
       "      <td></td>\n",
       "      <td>MIT9313</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>RFAM:RF00059</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS00920</td>\n",
       "      <td></td>\n",
       "      <td>rRNA</td>\n",
       "      <td>AKG35_RS00920</td>\n",
       "      <td>+</td>\n",
       "      <td>208846</td>\n",
       "      <td>210328</td>\n",
       "      <td>208846</td>\n",
       "      <td>210328</td>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS00925,RNA_53</td>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS00925</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS00925</td>\n",
       "      <td>+</td>\n",
       "      <td>210510</td>\n",
       "      <td>210584</td>\n",
       "      <td>210510</td>\n",
       "      <td>210584</td>\n",
       "      <td>tRNA-Ile</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS00930,RNA_2</td>\n",
       "      <td>tRNA-Ile</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS00930</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS00930</td>\n",
       "      <td>+</td>\n",
       "      <td>210593</td>\n",
       "      <td>210666</td>\n",
       "      <td>210593</td>\n",
       "      <td>210666</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS00935,RNA_3</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS00940</td>\n",
       "      <td></td>\n",
       "      <td>rRNA</td>\n",
       "      <td>AKG35_RS00940</td>\n",
       "      <td>+</td>\n",
       "      <td>211155</td>\n",
       "      <td>214034</td>\n",
       "      <td>211155</td>\n",
       "      <td>214034</td>\n",
       "      <td>23S ribosomal RNA</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS00945,RNA_54</td>\n",
       "      <td>23S ribosomal RNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS00945</td>\n",
       "      <td></td>\n",
       "      <td>rRNA</td>\n",
       "      <td>AKG35_RS00945</td>\n",
       "      <td>+</td>\n",
       "      <td>214141</td>\n",
       "      <td>214258</td>\n",
       "      <td>214141</td>\n",
       "      <td>214258</td>\n",
       "      <td>5S ribosomal RNA</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS00950,RNA_55</td>\n",
       "      <td>5S ribosomal RNA</td>\n",
       "      <td>RFAM:RF00001</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS01015</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS01015</td>\n",
       "      <td>-</td>\n",
       "      <td>227744</td>\n",
       "      <td>227816</td>\n",
       "      <td>227744</td>\n",
       "      <td>227816</td>\n",
       "      <td>tRNA-Thr</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS01020,RNA_43</td>\n",
       "      <td>tRNA-Thr</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS01020</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS01020</td>\n",
       "      <td>-</td>\n",
       "      <td>227826</td>\n",
       "      <td>227908</td>\n",
       "      <td>227826</td>\n",
       "      <td>227908</td>\n",
       "      <td>tRNA-Tyr</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS01025,RNA_42</td>\n",
       "      <td>tRNA-Tyr</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>regulatory</td>\n",
       "      <td></td>\n",
       "      <td>-</td>\n",
       "      <td>267198</td>\n",
       "      <td>267357</td>\n",
       "      <td>267198</td>\n",
       "      <td>267357</td>\n",
       "      <td></td>\n",
       "      <td>MIT9313</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>RFAM:RF00174</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS01535</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS01535</td>\n",
       "      <td>+</td>\n",
       "      <td>329017</td>\n",
       "      <td>329102</td>\n",
       "      <td>329017</td>\n",
       "      <td>329102</td>\n",
       "      <td>tRNA-Ser</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS01535,RNA_4</td>\n",
       "      <td>tRNA-Ser</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS12115</td>\n",
       "      <td></td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>AKG35_RS12115</td>\n",
       "      <td>+</td>\n",
       "      <td>365280</td>\n",
       "      <td>365377</td>\n",
       "      <td>365280</td>\n",
       "      <td>365377</td>\n",
       "      <td>signal recognition particle sRNA small type</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td></td>\n",
       "      <td>signal recognition particle sRNA small type</td>\n",
       "      <td>RFAM:RF00169</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS01730</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS01730</td>\n",
       "      <td>-</td>\n",
       "      <td>370809</td>\n",
       "      <td>370882</td>\n",
       "      <td>370809</td>\n",
       "      <td>370882</td>\n",
       "      <td>tRNA-His</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS01735,RNA_41</td>\n",
       "      <td>tRNA-His</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS01765</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS01765</td>\n",
       "      <td>-</td>\n",
       "      <td>379359</td>\n",
       "      <td>379441</td>\n",
       "      <td>379359</td>\n",
       "      <td>379441</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS01775,RNA_40</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS01775</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS01775</td>\n",
       "      <td>+</td>\n",
       "      <td>380927</td>\n",
       "      <td>381011</td>\n",
       "      <td>380927</td>\n",
       "      <td>381011</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS01785,RNA_5</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS02030</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS02030</td>\n",
       "      <td>+</td>\n",
       "      <td>430627</td>\n",
       "      <td>430700</td>\n",
       "      <td>430627</td>\n",
       "      <td>430700</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS02040,RNA_6</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS02470</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS02470</td>\n",
       "      <td>+</td>\n",
       "      <td>518779</td>\n",
       "      <td>518865</td>\n",
       "      <td>518779</td>\n",
       "      <td>518865</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS02480,RNA_7</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS02580</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>AKG35_RS02580</td>\n",
       "      <td>-</td>\n",
       "      <td>540452</td>\n",
       "      <td>540526</td>\n",
       "      <td>540452</td>\n",
       "      <td>540526</td>\n",
       "      <td>tRNA-Pro</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>RG24_RS02590,RNA_39</td>\n",
       "      <td>tRNA-Pro</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS12245</td>\n",
       "      <td>PMT_ncRNA_Yfr7</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>AKG35_RS12245</td>\n",
       "      <td>+</td>\n",
       "      <td>727039</td>\n",
       "      <td>727223</td>\n",
       "      <td>727039</td>\n",
       "      <td>727223</td>\n",
       "      <td>6S RNA</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>PMT_ncRNA_Yfr7</td>\n",
       "      <td>6S RNA</td>\n",
       "      <td>RFAM:RF00013</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    contig_id        gene_id          pmt_id        type       location  \\\n",
       "44    MIT9313  AKG35_RS00225                        tRNA  AKG35_RS00225   \n",
       "84    MIT9313  AKG35_RS00430                        tRNA  AKG35_RS00430   \n",
       "162   MIT9313                                 regulatory                  \n",
       "187   MIT9313  AKG35_RS00920                        rRNA  AKG35_RS00920   \n",
       "188   MIT9313  AKG35_RS00925                        tRNA  AKG35_RS00925   \n",
       "189   MIT9313  AKG35_RS00930                        tRNA  AKG35_RS00930   \n",
       "190   MIT9313  AKG35_RS00940                        rRNA  AKG35_RS00940   \n",
       "191   MIT9313  AKG35_RS00945                        rRNA  AKG35_RS00945   \n",
       "207   MIT9313  AKG35_RS01015                        tRNA  AKG35_RS01015   \n",
       "208   MIT9313  AKG35_RS01020                        tRNA  AKG35_RS01020   \n",
       "250   MIT9313                                 regulatory                  \n",
       "312   MIT9313  AKG35_RS01535                        tRNA  AKG35_RS01535   \n",
       "345   MIT9313  AKG35_RS12115                       ncRNA  AKG35_RS12115   \n",
       "351   MIT9313  AKG35_RS01730                        tRNA  AKG35_RS01730   \n",
       "359   MIT9313  AKG35_RS01765                        tRNA  AKG35_RS01765   \n",
       "361   MIT9313  AKG35_RS01775                        tRNA  AKG35_RS01775   \n",
       "411   MIT9313  AKG35_RS02030                        tRNA  AKG35_RS02030   \n",
       "498   MIT9313  AKG35_RS02470                        tRNA  AKG35_RS02470   \n",
       "520   MIT9313  AKG35_RS02580                        tRNA  AKG35_RS02580   \n",
       "702   MIT9313  AKG35_RS12245  PMT_ncRNA_Yfr7       ncRNA  AKG35_RS12245   \n",
       "\n",
       "    strand   start    stop    left   right  \\\n",
       "44       -   42281   42353   42281   42353   \n",
       "84       +   87819   87891   87819   87891   \n",
       "162      +  178355  178459  178355  178459   \n",
       "187      +  208846  210328  208846  210328   \n",
       "188      +  210510  210584  210510  210584   \n",
       "189      +  210593  210666  210593  210666   \n",
       "190      +  211155  214034  211155  214034   \n",
       "191      +  214141  214258  214141  214258   \n",
       "207      -  227744  227816  227744  227816   \n",
       "208      -  227826  227908  227826  227908   \n",
       "250      -  267198  267357  267198  267357   \n",
       "312      +  329017  329102  329017  329102   \n",
       "345      +  365280  365377  365280  365377   \n",
       "351      -  370809  370882  370809  370882   \n",
       "359      -  379359  379441  379359  379441   \n",
       "361      +  380927  381011  380927  381011   \n",
       "411      +  430627  430700  430627  430700   \n",
       "498      +  518779  518865  518779  518865   \n",
       "520      -  540452  540526  540452  540526   \n",
       "702      +  727039  727223  727039  727223   \n",
       "\n",
       "                                        function   genome  \\\n",
       "44                                      tRNA-Gly  MIT9313   \n",
       "84                                      tRNA-Gly  MIT9313   \n",
       "162                                               MIT9313   \n",
       "187                            16S ribosomal RNA  MIT9313   \n",
       "188                                     tRNA-Ile  MIT9313   \n",
       "189                                     tRNA-Ala  MIT9313   \n",
       "190                            23S ribosomal RNA  MIT9313   \n",
       "191                             5S ribosomal RNA  MIT9313   \n",
       "207                                     tRNA-Thr  MIT9313   \n",
       "208                                     tRNA-Tyr  MIT9313   \n",
       "250                                               MIT9313   \n",
       "312                                     tRNA-Ser  MIT9313   \n",
       "345  signal recognition particle sRNA small type  MIT9313   \n",
       "351                                     tRNA-His  MIT9313   \n",
       "359                                     tRNA-Leu  MIT9313   \n",
       "361                                     tRNA-Leu  MIT9313   \n",
       "411                                     tRNA-Ala  MIT9313   \n",
       "498                                     tRNA-Leu  MIT9313   \n",
       "520                                     tRNA-Pro  MIT9313   \n",
       "702                                       6S RNA  MIT9313   \n",
       "\n",
       "           old_locus_tag                                      product  \\\n",
       "44   RG24_RS00225,RNA_44                                     tRNA-Gly   \n",
       "84    RG24_RS00430,RNA_1                                     tRNA-Gly   \n",
       "162                                                                     \n",
       "187  RG24_RS00925,RNA_53                            16S ribosomal RNA   \n",
       "188   RG24_RS00930,RNA_2                                     tRNA-Ile   \n",
       "189   RG24_RS00935,RNA_3                                     tRNA-Ala   \n",
       "190  RG24_RS00945,RNA_54                            23S ribosomal RNA   \n",
       "191  RG24_RS00950,RNA_55                             5S ribosomal RNA   \n",
       "207  RG24_RS01020,RNA_43                                     tRNA-Thr   \n",
       "208  RG24_RS01025,RNA_42                                     tRNA-Tyr   \n",
       "250                                                                     \n",
       "312   RG24_RS01535,RNA_4                                     tRNA-Ser   \n",
       "345                       signal recognition particle sRNA small type   \n",
       "351  RG24_RS01735,RNA_41                                     tRNA-His   \n",
       "359  RG24_RS01775,RNA_40                                     tRNA-Leu   \n",
       "361   RG24_RS01785,RNA_5                                     tRNA-Leu   \n",
       "411   RG24_RS02040,RNA_6                                     tRNA-Ala   \n",
       "498   RG24_RS02480,RNA_7                                     tRNA-Leu   \n",
       "520  RG24_RS02590,RNA_39                                     tRNA-Pro   \n",
       "702       PMT_ncRNA_Yfr7                                       6S RNA   \n",
       "\n",
       "          db_xref protein_id figfam nucleotide_sequence aa_sequence  \n",
       "44                                                                   \n",
       "84                                                                   \n",
       "162  RFAM:RF00059                                                    \n",
       "187                                                                  \n",
       "188                                                                  \n",
       "189                                                                  \n",
       "190                                                                  \n",
       "191  RFAM:RF00001                                                    \n",
       "207                                                                  \n",
       "208                                                                  \n",
       "250  RFAM:RF00174                                                    \n",
       "312                                                                  \n",
       "345  RFAM:RF00169                                                    \n",
       "351                                                                  \n",
       "359                                                                  \n",
       "361                                                                  \n",
       "411                                                                  \n",
       "498                                                                  \n",
       "520                                                                  \n",
       "702  RFAM:RF00013                                                    "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.loc[gdf.type != 'CDS'].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'BX548175'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome = 'MIT9313'\n",
    "accession = sf.get_accession(genome)\n",
    "accession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "orf_dpath = 'orffinder_tmp'\n",
    "orf_fpath = os.path.join(orf_dpath, f'{accession}.orffinder.fa')\n",
    "df = sf.parse_orffinder_fasta(orf_fpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['rast_left'] = df['left']\n",
    "df['rast_right'] = df['right'] + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>pmt_id</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>function</th>\n",
       "      <th>genome</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>figfam</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIT9313</td>\n",
       "      <td>AKG35_RS00020</td>\n",
       "      <td>PMT0004</td>\n",
       "      <td>CDS</td>\n",
       "      <td>AKG35_RS00020</td>\n",
       "      <td>+</td>\n",
       "      <td>4621</td>\n",
       "      <td>6079</td>\n",
       "      <td>4621</td>\n",
       "      <td>6079</td>\n",
       "      <td>amidophosphoribosyltransferase</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>PMT0004,RG24_RS00020</td>\n",
       "      <td>amidophosphoribosyltransferase</td>\n",
       "      <td>GI:499441919</td>\n",
       "      <td>WP_011129383.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MCGIVGIVSTALVNQQIYDSLLLLQHRGQDSTGIATMDGSVFHLHK...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  contig_id        gene_id   pmt_id type       location strand  start  stop  \\\n",
       "3   MIT9313  AKG35_RS00020  PMT0004  CDS  AKG35_RS00020      +   4621  6079   \n",
       "\n",
       "   left  right                        function   genome         old_locus_tag  \\\n",
       "3  4621   6079  amidophosphoribosyltransferase  MIT9313  PMT0004,RG24_RS00020   \n",
       "\n",
       "                          product       db_xref      protein_id figfam  \\\n",
       "3  amidophosphoribosyltransferase  GI:499441919  WP_011129383.1          \n",
       "\n",
       "  nucleotide_sequence                                        aa_sequence  \n",
       "3                      MCGIVGIVSTALVNQQIYDSLLLLQHRGQDSTGIATMDGSVFHLHK...  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.loc[gdf.location == 'AKG35_RS00020']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>5047</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>orfid</th>\n",
       "      <td>ORF5048_BX548175.1:4621:6078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>contig</th>\n",
       "      <td>BX548175.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>left</th>\n",
       "      <td>4621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>right</th>\n",
       "      <td>6078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start</th>\n",
       "      <td>4621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop</th>\n",
       "      <td>6078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strand</th>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaseq</th>\n",
       "      <td>MCGIVGIVSTALVNQQIYDSLLLLQHRGQDSTGIATMDGSVFHLHK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>len_nn</th>\n",
       "      <td>1458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>len_aa</th>\n",
       "      <td>485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>len_aa3</th>\n",
       "      <td>1458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rast_left</th>\n",
       "      <td>4621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rast_right</th>\n",
       "      <td>6079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>contig_id</th>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <td>AKG35_RS00020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>location</th>\n",
       "      <td>AKG35_RS00020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>type</th>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_r</th>\n",
       "      <td>4621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_r</th>\n",
       "      <td>6079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>function</th>\n",
       "      <td>amidophosphoribosyltransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>figfam</th>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_sequence</th>\n",
       "      <td>MCGIVGIVSTALVNQQIYDSLLLLQHRGQDSTGIATMDGSVFHLHK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome</th>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>left_r</th>\n",
       "      <td>4621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>right_r</th>\n",
       "      <td>6079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>overlap_location</th>\n",
       "      <td>AKG35_RS00020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>overlap_strand</th>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>overlap_gene_type</th>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>overlap_count</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_same_strand</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_out_of_frame</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_inside</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_upstream</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_downstream</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>overlap_type</th>\n",
       "      <td>known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>otype</th>\n",
       "      <td>known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  5047\n",
       "orfid                                     ORF5048_BX548175.1:4621:6078\n",
       "contig                                                      BX548175.1\n",
       "left                                                              4621\n",
       "right                                                             6078\n",
       "start                                                             4621\n",
       "stop                                                              6078\n",
       "strand                                                               +\n",
       "aaseq                MCGIVGIVSTALVNQQIYDSLLLLQHRGQDSTGIATMDGSVFHLHK...\n",
       "len_nn                                                            1458\n",
       "len_aa                                                             485\n",
       "len_aa3                                                           1458\n",
       "rast_left                                                         4621\n",
       "rast_right                                                        6079\n",
       "contig_id                                                      MIT9313\n",
       "gene_id                                                  AKG35_RS00020\n",
       "location                                                 AKG35_RS00020\n",
       "type                                                               CDS\n",
       "start_r                                                           4621\n",
       "stop_r                                                            6079\n",
       "function                                amidophosphoribosyltransferase\n",
       "figfam                                                                \n",
       "nucleotide_sequence                                                   \n",
       "aa_sequence          MCGIVGIVSTALVNQQIYDSLLLLQHRGQDSTGIATMDGSVFHLHK...\n",
       "genome                                                         MIT9313\n",
       "left_r                                                            4621\n",
       "right_r                                                           6079\n",
       "overlap_location                                         AKG35_RS00020\n",
       "overlap_strand                                                       +\n",
       "overlap_gene_type                                                  CDS\n",
       "overlap_count                                                        1\n",
       "is_same_strand                                                     NaN\n",
       "is_out_of_frame                                                    NaN\n",
       "is_inside                                                          NaN\n",
       "is_upstream                                                        NaN\n",
       "is_downstream                                                      NaN\n",
       "overlap_type                                                     known\n",
       "otype                                                            known"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[(df.rast_left > 4620) & (df.rast_left < 4630),  ].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = sf.add_gene_annotation_to_sorf_df(df, gdf)\n",
    "df = sf.add_overlapping_genes(df, gdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_df = sf.gen_summary_df(df, genome)\n",
    "df.to_feather(os.path.join(orf_dpath, f'{genome}.ncbi.sorf.feather.gz'))\n",
    "summary_df.to_csv(os.path.join(orf_dpath, f'{genome}.ncbi.sorf.summary.csv.gz'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>otype</th>\n",
       "      <th>count</th>\n",
       "      <th>subset</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>antisense</td>\n",
       "      <td>13413</td>\n",
       "      <td>all</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>in_frame</td>\n",
       "      <td>1443</td>\n",
       "      <td>all</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>known</td>\n",
       "      <td>1472</td>\n",
       "      <td>all</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>out_frame</td>\n",
       "      <td>8841</td>\n",
       "      <td>all</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>overlap</td>\n",
       "      <td>1494</td>\n",
       "      <td>all</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>standalone</td>\n",
       "      <td>4667</td>\n",
       "      <td>all</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0</td>\n",
       "      <td>antisense</td>\n",
       "      <td>12985</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>in_frame</td>\n",
       "      <td>1209</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>known</td>\n",
       "      <td>300</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>out_frame</td>\n",
       "      <td>8504</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4</td>\n",
       "      <td>overlap</td>\n",
       "      <td>1105</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>5</td>\n",
       "      <td>standalone</td>\n",
       "      <td>4655</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0</td>\n",
       "      <td>antisense</td>\n",
       "      <td>10830</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>in_frame</td>\n",
       "      <td>1033</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2</td>\n",
       "      <td>known</td>\n",
       "      <td>30</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3</td>\n",
       "      <td>out_frame</td>\n",
       "      <td>7441</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>4</td>\n",
       "      <td>overlap</td>\n",
       "      <td>696</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>5</td>\n",
       "      <td>standalone</td>\n",
       "      <td>4336</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index       otype  count  subset   genome\n",
       "0       0   antisense  13413     all  MIT9313\n",
       "1       1    in_frame   1443     all  MIT9313\n",
       "2       2       known   1472     all  MIT9313\n",
       "3       3   out_frame   8841     all  MIT9313\n",
       "4       4     overlap   1494     all  MIT9313\n",
       "5       5  standalone   4667     all  MIT9313\n",
       "6       0   antisense  12985  <100aa  MIT9313\n",
       "7       1    in_frame   1209  <100aa  MIT9313\n",
       "8       2       known    300  <100aa  MIT9313\n",
       "9       3   out_frame   8504  <100aa  MIT9313\n",
       "10      4     overlap   1105  <100aa  MIT9313\n",
       "11      5  standalone   4655  <100aa  MIT9313\n",
       "12      0   antisense  10830   <50aa  MIT9313\n",
       "13      1    in_frame   1033   <50aa  MIT9313\n",
       "14      2       known     30   <50aa  MIT9313\n",
       "15      3   out_frame   7441   <50aa  MIT9313\n",
       "16      4     overlap    696   <50aa  MIT9313\n",
       "17      5  standalone   4336   <50aa  MIT9313"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
