{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import sys, os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.special import comb\n",
    "from scipy import stats\n",
    "import scipy.cluster.hierarchy as hac\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('classic')\n",
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from collections import Counter\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "sns.set(rc={'figure.figsize':(15,8)})\n",
    "sns.set_context('poster')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "#import CCPA_lib as cp\n",
    "import sorf_lib as sf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context('poster')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(Prochlorococcus marinus str. MIT 9313, complete genome\n",
    "https://www.ncbi.nlm.nih.gov/nuccore/NC_005071.1/)\n",
    "\n",
    " \n",
    "\n",
    "Prochlorococcus marinus MED4 complete genome\n",
    "https://www.ncbi.nlm.nih.gov/nuccore/BX548174.1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "![](http://oregonstate.edu/instruct/bb450/fall14/stryer7/2/table_02_02.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.uniprot.org/uniprot/Q7V735\n",
    "\n",
    "http://tigrfams.jcvi.org/cgi-bin/HmmReportPage.cgi?acc=TIGR03798\n",
    "\n",
    "https://www.ebi.ac.uk/training/online/course/interpro-functional-and-structural-analysis-protei/sequence-searching/searching-interpro-batc\n",
    "\n",
    "http://www.ebi.ac.uk/interpro/sequencesearch/iprscan5-S20190707-131508-0462-76111813-p1m\n",
    "\n",
    "https://www.ebi.ac.uk/Tools/services/rest/iprscan5/result/iprscan5-S20190707-131508-0462-76111813-p1m/json"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "6 frame translation\n",
    "classify ORFs into:\n",
    "* known gene\n",
    "* inside gene (part of gene/antisense)\n",
    "* extending gene\n",
    "* outside of gene (independent/operon?)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome='MED4'"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "genome_df = sf.load_genome('MED4')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "genome_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get all sequence records for the specified genbank file\n",
    "recs = [rec for rec in SeqIO.parse(os.path.join('data', \"MED4.gb\"), \"genbank\")]\n",
    "\n",
    "# print the number of sequence records that were extracted\n",
    "len(recs)\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "recs[0].features[1].location, recs[0].features[2].qualifiers, recs[0].features[2].strand"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "recs[0].features[1].location, recs[0].features[1].qualifiers, recs[0].features[2].strand"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "recs[0].features[4].location, recs[0].features[4].qualifiers, recs[0].features[4].strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1331"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int(recs[0].features[1].location.end)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "genome_df.columns"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "{i.type for i in recs[0].features}"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "i = recs[0].features[2]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "i.qualifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "    merge_cols = ['contig_id', 'gene_id', 'location', 'type', \n",
    "                  'start', 'stop', 'strand',\n",
    "                  'function', 'figfam',\n",
    "                  'nucleotide_sequence', 'aa_sequence', 'genome', 'left', 'right']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def gb_to_record(i, genome='MED4'):\n",
    "    q = i.qualifiers\n",
    "    g = lambda x : ','.join(q.get(x, []))\n",
    "    pmtid = [i for i in q.get('old_locus_tag',[]) if i.startswith('PMT')]\n",
    "    record = {\n",
    "        'contig_id' : genome, \n",
    "        'gene_id' : g('locus_tag'), \n",
    "        'pmt_id': ','.join(pmtid), \n",
    "        'type' : i.type, \n",
    "        'location' : g('locus_tag'),  \n",
    "        'strand' : '+' if i.location.strand == 1 else '-',\n",
    "        'start' : int(i.location.start),\n",
    "        'stop' : int(i.location.end),\n",
    "        'left' : int(i.location.start),\n",
    "        'right' : int(i.location.end),\n",
    "        'function' : g('product'), \n",
    "        'genome' : genome, \n",
    "        'old_locus_tag' : g('old_locus_tag'), \n",
    "        'product' : g('product'), \n",
    "        'db_xref' : g('db_xref'), \n",
    "        'protein_id' : g('protein_id'), \n",
    "        'figfam' : '',\n",
    "        'nucleotide_sequence' : '', \n",
    "        'aa_sequence' : g('translation'), \n",
    "    }\n",
    "    return record\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "types_to_collect = ['CDS', 'ncRNA', 'rRNA', 'regulatory', 'tRNA', 'tmRNA']\n",
    "records = [gb_to_record(i) for i in recs[0].features if i.type in types_to_collect]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = pd.DataFrame(records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "gdf.to_pickle(os.path.join('data', f'{genome}.ncbi.gb.pkl.gz'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>pmt_id</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>function</th>\n",
       "      <th>genome</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>figfam</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_37</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>RNA_37</td>\n",
       "      <td>-</td>\n",
       "      <td>65453</td>\n",
       "      <td>65535</td>\n",
       "      <td>65453</td>\n",
       "      <td>65535</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMED4_asRNA_00641</td>\n",
       "      <td></td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>PMED4_asRNA_00641</td>\n",
       "      <td>-</td>\n",
       "      <td>67628</td>\n",
       "      <td>67875</td>\n",
       "      <td>67628</td>\n",
       "      <td>67875</td>\n",
       "      <td></td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMED4_asRNA_00771</td>\n",
       "      <td></td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>PMED4_asRNA_00771</td>\n",
       "      <td>-</td>\n",
       "      <td>79766</td>\n",
       "      <td>79930</td>\n",
       "      <td>79766</td>\n",
       "      <td>79930</td>\n",
       "      <td></td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_1</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>RNA_1</td>\n",
       "      <td>+</td>\n",
       "      <td>122895</td>\n",
       "      <td>122967</td>\n",
       "      <td>122895</td>\n",
       "      <td>122967</td>\n",
       "      <td>tRNA-Asn</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>tRNA-Asn</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMED4_asRNA_02331</td>\n",
       "      <td></td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>PMED4_asRNA_02331</td>\n",
       "      <td>+</td>\n",
       "      <td>223721</td>\n",
       "      <td>223828</td>\n",
       "      <td>223721</td>\n",
       "      <td>223828</td>\n",
       "      <td></td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_2</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>RNA_2</td>\n",
       "      <td>+</td>\n",
       "      <td>229472</td>\n",
       "      <td>229554</td>\n",
       "      <td>229472</td>\n",
       "      <td>229554</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_36</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>RNA_36</td>\n",
       "      <td>-</td>\n",
       "      <td>249344</td>\n",
       "      <td>249418</td>\n",
       "      <td>249344</td>\n",
       "      <td>249418</td>\n",
       "      <td>tRNA-Arg</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>tRNA-Arg</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMED4_asRNA_02731</td>\n",
       "      <td></td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>PMED4_asRNA_02731</td>\n",
       "      <td>+</td>\n",
       "      <td>258537</td>\n",
       "      <td>258669</td>\n",
       "      <td>258537</td>\n",
       "      <td>258669</td>\n",
       "      <td></td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_35</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>RNA_35</td>\n",
       "      <td>-</td>\n",
       "      <td>264441</td>\n",
       "      <td>264517</td>\n",
       "      <td>264441</td>\n",
       "      <td>264517</td>\n",
       "      <td>tRNA-Met</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>tRNA-Met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_34</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>RNA_34</td>\n",
       "      <td>-</td>\n",
       "      <td>296376</td>\n",
       "      <td>296452</td>\n",
       "      <td>296376</td>\n",
       "      <td>296452</td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMED4_asRNA_03211</td>\n",
       "      <td></td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>PMED4_asRNA_03211</td>\n",
       "      <td>+</td>\n",
       "      <td>302064</td>\n",
       "      <td>302662</td>\n",
       "      <td>302064</td>\n",
       "      <td>302662</td>\n",
       "      <td></td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_33</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>RNA_33</td>\n",
       "      <td>-</td>\n",
       "      <td>305095</td>\n",
       "      <td>305167</td>\n",
       "      <td>305095</td>\n",
       "      <td>305167</td>\n",
       "      <td>tRNA-Thr</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>tRNA-Thr</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_38</td>\n",
       "      <td></td>\n",
       "      <td>rRNA</td>\n",
       "      <td>RNA_38</td>\n",
       "      <td>+</td>\n",
       "      <td>313060</td>\n",
       "      <td>318126</td>\n",
       "      <td>313060</td>\n",
       "      <td>318126</td>\n",
       "      <td>rRNA operon</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>rRNA operon</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_39</td>\n",
       "      <td></td>\n",
       "      <td>rRNA</td>\n",
       "      <td>RNA_39</td>\n",
       "      <td>+</td>\n",
       "      <td>313060</td>\n",
       "      <td>314525</td>\n",
       "      <td>313060</td>\n",
       "      <td>314525</td>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMED4_asRNA_38</td>\n",
       "      <td></td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>PMED4_asRNA_38</td>\n",
       "      <td>-</td>\n",
       "      <td>314573</td>\n",
       "      <td>314884</td>\n",
       "      <td>314573</td>\n",
       "      <td>314884</td>\n",
       "      <td></td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_3</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>RNA_3</td>\n",
       "      <td>+</td>\n",
       "      <td>314648</td>\n",
       "      <td>314722</td>\n",
       "      <td>314648</td>\n",
       "      <td>314722</td>\n",
       "      <td>tRNA-Ile</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>tRNA-Ile</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_4</td>\n",
       "      <td></td>\n",
       "      <td>tRNA</td>\n",
       "      <td>RNA_4</td>\n",
       "      <td>+</td>\n",
       "      <td>314734</td>\n",
       "      <td>314807</td>\n",
       "      <td>314734</td>\n",
       "      <td>314807</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_40</td>\n",
       "      <td></td>\n",
       "      <td>rRNA</td>\n",
       "      <td>RNA_40</td>\n",
       "      <td>+</td>\n",
       "      <td>315073</td>\n",
       "      <td>317948</td>\n",
       "      <td>315073</td>\n",
       "      <td>317948</td>\n",
       "      <td>23S ribosomal RNA</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>23S ribosomal RNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>MED4</td>\n",
       "      <td>RNA_41</td>\n",
       "      <td></td>\n",
       "      <td>rRNA</td>\n",
       "      <td>RNA_41</td>\n",
       "      <td>+</td>\n",
       "      <td>318011</td>\n",
       "      <td>318126</td>\n",
       "      <td>318011</td>\n",
       "      <td>318126</td>\n",
       "      <td>5S ribosomal RNA</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>5S ribosomal RNA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMED4_asRNA_03431</td>\n",
       "      <td></td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>PMED4_asRNA_03431</td>\n",
       "      <td>+</td>\n",
       "      <td>324059</td>\n",
       "      <td>324207</td>\n",
       "      <td>324059</td>\n",
       "      <td>324207</td>\n",
       "      <td></td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    contig_id            gene_id pmt_id   type           location strand  \\\n",
       "63       MED4             RNA_37          tRNA             RNA_37      -   \n",
       "66       MED4  PMED4_asRNA_00641         ncRNA  PMED4_asRNA_00641      -   \n",
       "80       MED4  PMED4_asRNA_00771         ncRNA  PMED4_asRNA_00771      -   \n",
       "135      MED4              RNA_1          tRNA              RNA_1      +   \n",
       "239      MED4  PMED4_asRNA_02331         ncRNA  PMED4_asRNA_02331      +   \n",
       "249      MED4              RNA_2          tRNA              RNA_2      +   \n",
       "272      MED4             RNA_36          tRNA             RNA_36      -   \n",
       "283      MED4  PMED4_asRNA_02731         ncRNA  PMED4_asRNA_02731      +   \n",
       "290      MED4             RNA_35          tRNA             RNA_35      -   \n",
       "327      MED4             RNA_34          tRNA             RNA_34      -   \n",
       "334      MED4  PMED4_asRNA_03211         ncRNA  PMED4_asRNA_03211      +   \n",
       "340      MED4             RNA_33          tRNA             RNA_33      -   \n",
       "349      MED4             RNA_38          rRNA             RNA_38      +   \n",
       "350      MED4             RNA_39          rRNA             RNA_39      +   \n",
       "351      MED4     PMED4_asRNA_38         ncRNA     PMED4_asRNA_38      -   \n",
       "352      MED4              RNA_3          tRNA              RNA_3      +   \n",
       "353      MED4              RNA_4          tRNA              RNA_4      +   \n",
       "354      MED4             RNA_40          rRNA             RNA_40      +   \n",
       "355      MED4             RNA_41          rRNA             RNA_41      +   \n",
       "364      MED4  PMED4_asRNA_03431         ncRNA  PMED4_asRNA_03431      +   \n",
       "\n",
       "      start    stop    left   right           function genome old_locus_tag  \\\n",
       "63    65453   65535   65453   65535           tRNA-Leu   MED4                 \n",
       "66    67628   67875   67628   67875                      MED4                 \n",
       "80    79766   79930   79766   79930                      MED4                 \n",
       "135  122895  122967  122895  122967           tRNA-Asn   MED4                 \n",
       "239  223721  223828  223721  223828                      MED4                 \n",
       "249  229472  229554  229472  229554           tRNA-Leu   MED4                 \n",
       "272  249344  249418  249344  249418           tRNA-Arg   MED4                 \n",
       "283  258537  258669  258537  258669                      MED4                 \n",
       "290  264441  264517  264441  264517           tRNA-Met   MED4                 \n",
       "327  296376  296452  296376  296452           tRNA-Phe   MED4                 \n",
       "334  302064  302662  302064  302662                      MED4                 \n",
       "340  305095  305167  305095  305167           tRNA-Thr   MED4                 \n",
       "349  313060  318126  313060  318126        rRNA operon   MED4                 \n",
       "350  313060  314525  313060  314525  16S ribosomal RNA   MED4                 \n",
       "351  314573  314884  314573  314884                      MED4                 \n",
       "352  314648  314722  314648  314722           tRNA-Ile   MED4                 \n",
       "353  314734  314807  314734  314807           tRNA-Ala   MED4                 \n",
       "354  315073  317948  315073  317948  23S ribosomal RNA   MED4                 \n",
       "355  318011  318126  318011  318126   5S ribosomal RNA   MED4                 \n",
       "364  324059  324207  324059  324207                      MED4                 \n",
       "\n",
       "               product db_xref protein_id figfam nucleotide_sequence  \\\n",
       "63            tRNA-Leu                                                 \n",
       "66                                                                     \n",
       "80                                                                     \n",
       "135           tRNA-Asn                                                 \n",
       "239                                                                    \n",
       "249           tRNA-Leu                                                 \n",
       "272           tRNA-Arg                                                 \n",
       "283                                                                    \n",
       "290           tRNA-Met                                                 \n",
       "327           tRNA-Phe                                                 \n",
       "334                                                                    \n",
       "340           tRNA-Thr                                                 \n",
       "349        rRNA operon                                                 \n",
       "350  16S ribosomal RNA                                                 \n",
       "351                                                                    \n",
       "352           tRNA-Ile                                                 \n",
       "353           tRNA-Ala                                                 \n",
       "354  23S ribosomal RNA                                                 \n",
       "355   5S ribosomal RNA                                                 \n",
       "364                                                                    \n",
       "\n",
       "    aa_sequence  \n",
       "63               \n",
       "66               \n",
       "80               \n",
       "135              \n",
       "239              \n",
       "249              \n",
       "272              \n",
       "283              \n",
       "290              \n",
       "327              \n",
       "334              \n",
       "340              \n",
       "349              \n",
       "350              \n",
       "351              \n",
       "352              \n",
       "353              \n",
       "354              \n",
       "355              \n",
       "364              "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.loc[gdf.type != 'CDS'].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'BX548174'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome = 'MED4'\n",
    "accession = sf.get_accession(genome)\n",
    "accession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "orf_dpath = 'orffinder_tmp'\n",
    "orf_fpath = os.path.join(orf_dpath, f'{accession}.orffinder.fa')\n",
    "df = sf.parse_orffinder_fasta(orf_fpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['rast_left'] = df['left']\n",
    "df['rast_right'] = df['right'] + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>pmt_id</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>function</th>\n",
       "      <th>genome</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>figfam</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [contig_id, gene_id, pmt_id, type, location, strand, start, stop, left, right, function, genome, old_locus_tag, product, db_xref, protein_id, figfam, nucleotide_sequence, aa_sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.loc[gdf.location == 'AKG35_RS00020']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>orfid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>contig</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>left</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>right</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaseq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>len_nn</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>len_aa</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>len_aa3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rast_left</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rast_right</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [orfid, contig, left, right, start, stop, strand, aaseq, len_nn, len_aa, len_aa3, rast_left, rast_right]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[(df.rast_left > 4620) & (df.rast_left < 4630),  ].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = sf.add_gene_annotation_to_sorf_df(df, gdf)\n",
    "df = sf.add_overlapping_genes(df, gdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_df = sf.gen_summary_df(df, genome)\n",
    "df.to_pickle(os.path.join(orf_dpath, f'{genome}.ncbi.sorf.pkl.gz'))\n",
    "summary_df.to_csv(os.path.join(orf_dpath, f'{genome}.ncbi.sorf.summary.csv.gz'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>otype</th>\n",
       "      <th>count</th>\n",
       "      <th>subset</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>antisense</td>\n",
       "      <td>7381</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>in_frame</td>\n",
       "      <td>931</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>known</td>\n",
       "      <td>1495</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>out_frame</td>\n",
       "      <td>5715</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>overlap</td>\n",
       "      <td>749</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>standalone</td>\n",
       "      <td>798</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0</td>\n",
       "      <td>antisense</td>\n",
       "      <td>7340</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>in_frame</td>\n",
       "      <td>727</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>known</td>\n",
       "      <td>395</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>out_frame</td>\n",
       "      <td>5714</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4</td>\n",
       "      <td>overlap</td>\n",
       "      <td>655</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>5</td>\n",
       "      <td>standalone</td>\n",
       "      <td>798</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0</td>\n",
       "      <td>antisense</td>\n",
       "      <td>6848</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>in_frame</td>\n",
       "      <td>660</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2</td>\n",
       "      <td>known</td>\n",
       "      <td>97</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3</td>\n",
       "      <td>out_frame</td>\n",
       "      <td>5635</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>4</td>\n",
       "      <td>overlap</td>\n",
       "      <td>567</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>5</td>\n",
       "      <td>standalone</td>\n",
       "      <td>790</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index       otype  count  subset genome\n",
       "0       0   antisense   7381     all   MED4\n",
       "1       1    in_frame    931     all   MED4\n",
       "2       2       known   1495     all   MED4\n",
       "3       3   out_frame   5715     all   MED4\n",
       "4       4     overlap    749     all   MED4\n",
       "5       5  standalone    798     all   MED4\n",
       "6       0   antisense   7340  <100aa   MED4\n",
       "7       1    in_frame    727  <100aa   MED4\n",
       "8       2       known    395  <100aa   MED4\n",
       "9       3   out_frame   5714  <100aa   MED4\n",
       "10      4     overlap    655  <100aa   MED4\n",
       "11      5  standalone    798  <100aa   MED4\n",
       "12      0   antisense   6848   <50aa   MED4\n",
       "13      1    in_frame    660   <50aa   MED4\n",
       "14      2       known     97   <50aa   MED4\n",
       "15      3   out_frame   5635   <50aa   MED4\n",
       "16      4     overlap    567   <50aa   MED4\n",
       "17      5  standalone    790   <50aa   MED4"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "orfid = 'ORF16879_BX548175.1:1658121:1657948'\n",
    "orfid = 'ORF19962_BX548175.1:317208:317134'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>orfid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>contig</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>left</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>right</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aaseq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>len_nn</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>len_aa</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>len_aa3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rast_left</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rast_right</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>contig_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_r</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_r</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>function</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>figfam</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleotide_sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>left_r</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>right_r</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>overlap_location</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>overlap_strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>overlap_gene_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>overlap_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_same_strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_out_of_frame</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_inside</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_upstream</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_downstream</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>overlap_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>otype</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [orfid, contig, left, right, start, stop, strand, aaseq, len_nn, len_aa, len_aa3, rast_left, rast_right, contig_id, gene_id, location, type, start_r, stop_r, function, figfam, nucleotide_sequence, aa_sequence, genome, left_r, right_r, overlap_location, overlap_strand, overlap_gene_type, overlap_count, is_same_strand, is_out_of_frame, is_inside, is_upstream, is_downstream, overlap_type, otype]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.orfid.isin([orfid])].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['contig_id', 'gene_id', 'pmt_id', 'type', 'location', 'strand', 'start',\n",
       "       'stop', 'left', 'right', 'function', 'genome', 'old_locus_tag',\n",
       "       'product', 'db_xref', 'protein_id', 'figfam', 'nucleotide_sequence',\n",
       "       'aa_sequence'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>pmt_id</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>function</th>\n",
       "      <th>genome</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>figfam</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [contig_id, gene_id, pmt_id, type, location, strand, start, stop, left, right, function, genome, old_locus_tag, product, db_xref, protein_id, figfam, nucleotide_sequence, aa_sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.loc[(gdf.left > 317100) & (gdf.right < 317220)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>pmt_id</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>function</th>\n",
       "      <th>genome</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>figfam</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM0002</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM0002</td>\n",
       "      <td>+</td>\n",
       "      <td>1332</td>\n",
       "      <td>2040</td>\n",
       "      <td>1332</td>\n",
       "      <td>2040</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM0002,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE18461.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM0006</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM0006</td>\n",
       "      <td>-</td>\n",
       "      <td>8415</td>\n",
       "      <td>9270</td>\n",
       "      <td>8415</td>\n",
       "      <td>9270</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM0006,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE18465.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM0007</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM0007</td>\n",
       "      <td>-</td>\n",
       "      <td>9274</td>\n",
       "      <td>10219</td>\n",
       "      <td>9274</td>\n",
       "      <td>10219</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM0007,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE18466.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM0008</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM0008</td>\n",
       "      <td>+</td>\n",
       "      <td>10366</td>\n",
       "      <td>11104</td>\n",
       "      <td>10366</td>\n",
       "      <td>11104</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM0008,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE18467.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM0014</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM0014</td>\n",
       "      <td>-</td>\n",
       "      <td>16679</td>\n",
       "      <td>17684</td>\n",
       "      <td>16679</td>\n",
       "      <td>17684</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM0014,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE18473.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MKSYQINTINNLHKISVAPMMDCTDKHFRMIIRKISSKALLYTEMI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1987</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM1681</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM1681</td>\n",
       "      <td>-</td>\n",
       "      <td>1613534</td>\n",
       "      <td>1614146</td>\n",
       "      <td>1613534</td>\n",
       "      <td>1614146</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM1681,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE20140.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MTLIFENIYNLAYFFNTNIGIFAFILLYILIVLLILPASWLSLLSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2005</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM1698</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM1698</td>\n",
       "      <td>+</td>\n",
       "      <td>1633926</td>\n",
       "      <td>1634376</td>\n",
       "      <td>1633926</td>\n",
       "      <td>1634376</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM1698,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE20157.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MLTFINYQSKLFPTIEDVDRKSYGTLFYCLSLFILIYLYWEKDPTS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2015</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM2020</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM2020</td>\n",
       "      <td>-</td>\n",
       "      <td>1646198</td>\n",
       "      <td>1646384</td>\n",
       "      <td>1646198</td>\n",
       "      <td>1646384</td>\n",
       "      <td>Conserved hypothetical protein</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>Conserved hypothetical protein</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM2020,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAP16525.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MFDKENLKYFLIWPMSVLLAIFFKYYGFLKPDFLLINNYLVLLLVC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2016</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM1708</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM1708</td>\n",
       "      <td>+</td>\n",
       "      <td>1646438</td>\n",
       "      <td>1646717</td>\n",
       "      <td>1646438</td>\n",
       "      <td>1646717</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM1708,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE20167.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MDSNKLKIRLNEIAEVNPALTCYHREDPAPVLPLREEPDLLSWLED...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2023</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM1715</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM1715</td>\n",
       "      <td>+</td>\n",
       "      <td>1656135</td>\n",
       "      <td>1656711</td>\n",
       "      <td>1656135</td>\n",
       "      <td>1656711</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM1715,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE20174.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MKFSKSLFPRIFLILFINQIFFNVPKAQSAEKIKIIYSIFSRTVTV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>575 rows  19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     contig_id  gene_id pmt_id type location strand    start     stop  \\\n",
       "1         MED4  PMM0002         CDS  PMM0002      +     1332     2040   \n",
       "5         MED4  PMM0006         CDS  PMM0006      -     8415     9270   \n",
       "6         MED4  PMM0007         CDS  PMM0007      -     9274    10219   \n",
       "7         MED4  PMM0008         CDS  PMM0008      +    10366    11104   \n",
       "13        MED4  PMM0014         CDS  PMM0014      -    16679    17684   \n",
       "...        ...      ...    ...  ...      ...    ...      ...      ...   \n",
       "1987      MED4  PMM1681         CDS  PMM1681      -  1613534  1614146   \n",
       "2005      MED4  PMM1698         CDS  PMM1698      +  1633926  1634376   \n",
       "2015      MED4  PMM2020         CDS  PMM2020      -  1646198  1646384   \n",
       "2016      MED4  PMM1708         CDS  PMM1708      +  1646438  1646717   \n",
       "2023      MED4  PMM1715         CDS  PMM1715      +  1656135  1656711   \n",
       "\n",
       "         left    right                        function genome old_locus_tag  \\\n",
       "1        1332     2040  conserved hypothetical protein   MED4                 \n",
       "5        8415     9270  conserved hypothetical protein   MED4                 \n",
       "6        9274    10219  conserved hypothetical protein   MED4                 \n",
       "7       10366    11104  conserved hypothetical protein   MED4                 \n",
       "13      16679    17684  conserved hypothetical protein   MED4                 \n",
       "...       ...      ...                             ...    ...           ...   \n",
       "1987  1613534  1614146  conserved hypothetical protein   MED4                 \n",
       "2005  1633926  1634376  conserved hypothetical protein   MED4                 \n",
       "2015  1646198  1646384  Conserved hypothetical protein   MED4                 \n",
       "2016  1646438  1646717  conserved hypothetical protein   MED4                 \n",
       "2023  1656135  1656711  conserved hypothetical protein   MED4                 \n",
       "\n",
       "                             product  \\\n",
       "1     conserved hypothetical protein   \n",
       "5     conserved hypothetical protein   \n",
       "6     conserved hypothetical protein   \n",
       "7     conserved hypothetical protein   \n",
       "13    conserved hypothetical protein   \n",
       "...                              ...   \n",
       "1987  conserved hypothetical protein   \n",
       "2005  conserved hypothetical protein   \n",
       "2015  Conserved hypothetical protein   \n",
       "2016  conserved hypothetical protein   \n",
       "2023  conserved hypothetical protein   \n",
       "\n",
       "                                                db_xref  protein_id figfam  \\\n",
       "1     EnsemblGenomes-Gn:PMM0002,EnsemblGenomes-Tr:CA...  CAE18461.1          \n",
       "5     EnsemblGenomes-Gn:PMM0006,EnsemblGenomes-Tr:CA...  CAE18465.1          \n",
       "6     EnsemblGenomes-Gn:PMM0007,EnsemblGenomes-Tr:CA...  CAE18466.1          \n",
       "7     EnsemblGenomes-Gn:PMM0008,EnsemblGenomes-Tr:CA...  CAE18467.1          \n",
       "13    EnsemblGenomes-Gn:PMM0014,EnsemblGenomes-Tr:CA...  CAE18473.1          \n",
       "...                                                 ...         ...    ...   \n",
       "1987  EnsemblGenomes-Gn:PMM1681,EnsemblGenomes-Tr:CA...  CAE20140.1          \n",
       "2005  EnsemblGenomes-Gn:PMM1698,EnsemblGenomes-Tr:CA...  CAE20157.1          \n",
       "2015  EnsemblGenomes-Gn:PMM2020,EnsemblGenomes-Tr:CA...  CAP16525.1          \n",
       "2016  EnsemblGenomes-Gn:PMM1708,EnsemblGenomes-Tr:CA...  CAE20167.1          \n",
       "2023  EnsemblGenomes-Gn:PMM1715,EnsemblGenomes-Tr:CA...  CAE20174.1          \n",
       "\n",
       "     nucleotide_sequence                                        aa_sequence  \n",
       "1                         MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...  \n",
       "5                         MKKSLLKILFFSIISSHVFIAESLKALIPYYYLPETKSLQKQGLSI...  \n",
       "6                         MRNMIQNKKEFSEKLKKRAIFEGFAVSGIASIPGSSRVKLRTQALE...  \n",
       "7                         MVESTQSQDSNLGTRLQQDLKNDLIAGLLVVIPLATTIWLSSIVSK...  \n",
       "13                        MKSYQINTINNLHKISVAPMMDCTDKHFRMIIRKISSKALLYTEMI...  \n",
       "...                  ...                                                ...  \n",
       "1987                      MTLIFENIYNLAYFFNTNIGIFAFILLYILIVLLILPASWLSLLSG...  \n",
       "2005                      MLTFINYQSKLFPTIEDVDRKSYGTLFYCLSLFILIYLYWEKDPTS...  \n",
       "2015                      MFDKENLKYFLIWPMSVLLAIFFKYYGFLKPDFLLINNYLVLLLVC...  \n",
       "2016                      MDSNKLKIRLNEIAEVNPALTCYHREDPAPVLPLREEPDLLSWLED...  \n",
       "2023                      MKFSKSLFPRIFLILFINQIFFNVPKAQSAEKIKIIYSIFSRTVTV...  \n",
       "\n",
       "[575 rows x 19 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.loc[gdf['product'].str.contains('hypothetical protein')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>pmt_id</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>function</th>\n",
       "      <th>genome</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>figfam</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [contig_id, gene_id, pmt_id, type, location, strand, start, stop, left, right, function, genome, old_locus_tag, product, db_xref, protein_id, figfam, nucleotide_sequence, aa_sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.loc[gdf.protein_id.str.contains('WP_011129965')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orfid</th>\n",
       "      <th>contig</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>aaseq</th>\n",
       "      <th>len_nn</th>\n",
       "      <th>len_aa</th>\n",
       "      <th>...</th>\n",
       "      <th>overlap_strand</th>\n",
       "      <th>overlap_gene_type</th>\n",
       "      <th>overlap_count</th>\n",
       "      <th>is_same_strand</th>\n",
       "      <th>is_out_of_frame</th>\n",
       "      <th>is_inside</th>\n",
       "      <th>is_upstream</th>\n",
       "      <th>is_downstream</th>\n",
       "      <th>overlap_type</th>\n",
       "      <th>otype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows  37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [orfid, contig, left, right, start, stop, strand, aaseq, len_nn, len_aa, len_aa3, rast_left, rast_right, contig_id, gene_id, location, type, start_r, stop_r, function, figfam, nucleotide_sequence, aa_sequence, genome, left_r, right_r, overlap_location, overlap_strand, overlap_gene_type, overlap_count, is_same_strand, is_out_of_frame, is_inside, is_upstream, is_downstream, overlap_type, otype]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 37 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.orfid== 'ORF30023_BX548175.1:644083:643928']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>otype</th>\n",
       "      <th>count</th>\n",
       "      <th>subset</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>antisense</td>\n",
       "      <td>7381</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>in_frame</td>\n",
       "      <td>931</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>known</td>\n",
       "      <td>1495</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>out_frame</td>\n",
       "      <td>5715</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>overlap</td>\n",
       "      <td>749</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>standalone</td>\n",
       "      <td>798</td>\n",
       "      <td>all</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0</td>\n",
       "      <td>antisense</td>\n",
       "      <td>7340</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>in_frame</td>\n",
       "      <td>727</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>known</td>\n",
       "      <td>395</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>out_frame</td>\n",
       "      <td>5714</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4</td>\n",
       "      <td>overlap</td>\n",
       "      <td>655</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>5</td>\n",
       "      <td>standalone</td>\n",
       "      <td>798</td>\n",
       "      <td>&lt;100aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0</td>\n",
       "      <td>antisense</td>\n",
       "      <td>6848</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>in_frame</td>\n",
       "      <td>660</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2</td>\n",
       "      <td>known</td>\n",
       "      <td>97</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3</td>\n",
       "      <td>out_frame</td>\n",
       "      <td>5635</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>4</td>\n",
       "      <td>overlap</td>\n",
       "      <td>567</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>5</td>\n",
       "      <td>standalone</td>\n",
       "      <td>790</td>\n",
       "      <td>&lt;50aa</td>\n",
       "      <td>MED4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index       otype  count  subset genome\n",
       "0       0   antisense   7381     all   MED4\n",
       "1       1    in_frame    931     all   MED4\n",
       "2       2       known   1495     all   MED4\n",
       "3       3   out_frame   5715     all   MED4\n",
       "4       4     overlap    749     all   MED4\n",
       "5       5  standalone    798     all   MED4\n",
       "6       0   antisense   7340  <100aa   MED4\n",
       "7       1    in_frame    727  <100aa   MED4\n",
       "8       2       known    395  <100aa   MED4\n",
       "9       3   out_frame   5714  <100aa   MED4\n",
       "10      4     overlap    655  <100aa   MED4\n",
       "11      5  standalone    798  <100aa   MED4\n",
       "12      0   antisense   6848   <50aa   MED4\n",
       "13      1    in_frame    660   <50aa   MED4\n",
       "14      2       known     97   <50aa   MED4\n",
       "15      3   out_frame   5635   <50aa   MED4\n",
       "16      4     overlap    567   <50aa   MED4\n",
       "17      5  standalone    790   <50aa   MED4"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "313060     2\n",
       "228280     2\n",
       "178177     1\n",
       "1397443    1\n",
       "1108672    1\n",
       "          ..\n",
       "750968     1\n",
       "1412467    1\n",
       "79218      1\n",
       "810353     1\n",
       "202751     1\n",
       "Name: left, Length: 2023, dtype: int64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.left.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>pmt_id</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "      <th>function</th>\n",
       "      <th>genome</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>figfam</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM0001</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM0001</td>\n",
       "      <td>+</td>\n",
       "      <td>173</td>\n",
       "      <td>1331</td>\n",
       "      <td>173</td>\n",
       "      <td>1331</td>\n",
       "      <td>DNA polymerase III, beta chain</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>DNA polymerase III, beta chain</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM0001,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE18460.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MEIVCNQNEFNYAIQLVSKAVASRPTHPILANLLLTADQGTNKISL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM0002</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM0002</td>\n",
       "      <td>+</td>\n",
       "      <td>1332</td>\n",
       "      <td>2040</td>\n",
       "      <td>1332</td>\n",
       "      <td>2040</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM0002,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE18461.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM0003</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM0003</td>\n",
       "      <td>+</td>\n",
       "      <td>2043</td>\n",
       "      <td>4383</td>\n",
       "      <td>2043</td>\n",
       "      <td>4383</td>\n",
       "      <td>phosphoribosylformylglycinamidine synthetase II</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>phosphoribosylformylglycinamidine synthetase II</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM0003,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE18462.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MIDSSSNNTYDVNESLKVENLTRDDYEEICKRLGRKPNRTELGMFG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM0004</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM0004</td>\n",
       "      <td>+</td>\n",
       "      <td>4429</td>\n",
       "      <td>5890</td>\n",
       "      <td>4429</td>\n",
       "      <td>5890</td>\n",
       "      <td>Glutamine amidotransferase class-II:Phosphorib...</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>Glutamine amidotransferase class-II:Phosphorib...</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM0004,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE18463.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MCGIVGIVSCDDVNQQIYDSLLLLQHRGQDSTGIATMENTVFHIHK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MED4</td>\n",
       "      <td>PMM0005</td>\n",
       "      <td></td>\n",
       "      <td>CDS</td>\n",
       "      <td>PMM0005</td>\n",
       "      <td>-</td>\n",
       "      <td>5886</td>\n",
       "      <td>8328</td>\n",
       "      <td>5886</td>\n",
       "      <td>8328</td>\n",
       "      <td>DNA gyrase/topoisomerase IV, subunit A</td>\n",
       "      <td>MED4</td>\n",
       "      <td></td>\n",
       "      <td>DNA gyrase/topoisomerase IV, subunit A</td>\n",
       "      <td>EnsemblGenomes-Gn:PMM0005,EnsemblGenomes-Tr:CA...</td>\n",
       "      <td>CAE18464.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MTKEKFTSISLQEEMQRSYLEYAMSVIIGRALPDARDGLKPVQRRI...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  contig_id  gene_id pmt_id type location strand  start  stop  left  right  \\\n",
       "0      MED4  PMM0001         CDS  PMM0001      +    173  1331   173   1331   \n",
       "1      MED4  PMM0002         CDS  PMM0002      +   1332  2040  1332   2040   \n",
       "2      MED4  PMM0003         CDS  PMM0003      +   2043  4383  2043   4383   \n",
       "3      MED4  PMM0004         CDS  PMM0004      +   4429  5890  4429   5890   \n",
       "4      MED4  PMM0005         CDS  PMM0005      -   5886  8328  5886   8328   \n",
       "\n",
       "                                            function genome old_locus_tag  \\\n",
       "0                     DNA polymerase III, beta chain   MED4                 \n",
       "1                     conserved hypothetical protein   MED4                 \n",
       "2    phosphoribosylformylglycinamidine synthetase II   MED4                 \n",
       "3  Glutamine amidotransferase class-II:Phosphorib...   MED4                 \n",
       "4             DNA gyrase/topoisomerase IV, subunit A   MED4                 \n",
       "\n",
       "                                             product  \\\n",
       "0                     DNA polymerase III, beta chain   \n",
       "1                     conserved hypothetical protein   \n",
       "2    phosphoribosylformylglycinamidine synthetase II   \n",
       "3  Glutamine amidotransferase class-II:Phosphorib...   \n",
       "4             DNA gyrase/topoisomerase IV, subunit A   \n",
       "\n",
       "                                             db_xref  protein_id figfam  \\\n",
       "0  EnsemblGenomes-Gn:PMM0001,EnsemblGenomes-Tr:CA...  CAE18460.1          \n",
       "1  EnsemblGenomes-Gn:PMM0002,EnsemblGenomes-Tr:CA...  CAE18461.1          \n",
       "2  EnsemblGenomes-Gn:PMM0003,EnsemblGenomes-Tr:CA...  CAE18462.1          \n",
       "3  EnsemblGenomes-Gn:PMM0004,EnsemblGenomes-Tr:CA...  CAE18463.1          \n",
       "4  EnsemblGenomes-Gn:PMM0005,EnsemblGenomes-Tr:CA...  CAE18464.1          \n",
       "\n",
       "  nucleotide_sequence                                        aa_sequence  \n",
       "0                      MEIVCNQNEFNYAIQLVSKAVASRPTHPILANLLLTADQGTNKISL...  \n",
       "1                      MKLPKEFLLSELLRHNVKGNATLNYGKGENVWMHPPVHRILGWYSR...  \n",
       "2                      MIDSSSNNTYDVNESLKVENLTRDDYEEICKRLGRKPNRTELGMFG...  \n",
       "3                      MCGIVGIVSCDDVNQQIYDSLLLLQHRGQDSTGIATMENTVFHIHK...  \n",
       "4                      MTKEKFTSISLQEEMQRSYLEYAMSVIIGRALPDARDGLKPVQRRI...  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CDS      1960\n",
       "tRNA       37\n",
       "ncRNA      24\n",
       "rRNA        4\n",
       "Name: type, dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf.type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
