{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import sys, os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.special import comb\n",
    "from scipy import stats\n",
    "import scipy.cluster.hierarchy as hac\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('classic')\n",
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "sns.set(rc={'figure.figsize':(15,8)})\n",
    "sns.set_context('poster')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "#import CCPA_lib as cp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context('poster')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Catalytic Promiscuity in the Biosynthesis of Cyclic Peptide in Planktonic Marine Cyanobacteria, PNAS2010\n",
    "Detection of procM and procA gene sequences in marine cyanobacterial genomes.\n",
    "1. **procM** gene sequences were detected in marine cyanobacterial genomes using the **pre-computed phylogenetic tree viewer in the microbesonline database** (http://www.microbesonline.org (3), which is computed using FastBLAST (4), see http://www.microbesonline.org/treebrowseHelp.html for additional details). \n",
    "2. **procA** sequences were detected in the genomes of Prochlorococcus **MIT9313** by **manually searching for short ORFs close to the procM gene** that encoded a **peptide** with a **leader sequence** followed by a **double-glycine** protease recognition site and a **short** core peptide rich in **Cys, Ser and Thr** residues (Fig. 1C). \n",
    "3. Once several such putative lantipeptides were discovered, we used **BLASTp** in both **microbesonline** and the **NCBI nr** database (with a cutoff **e-value of 10**) to detect other, similar genes in cyanobacterial genomes\n",
    "4.  **manually** checked the resulting peptide sequences to verify that they agreed with the definition of procAs, as described above. \n",
    "5. Figure 1 and Figure S1 were produced using IMG (http://img.jgi.doe.gov/cgi-bin/pub/main.cgi)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evolutionary radiation of lanthipeptides in marine cyanobacteria, PNAPLUS 2017\n",
    "1. Annotation of **prochlorosin** biosynthesis genes was performed manually using **BLAST** (12). \n",
    "2. The sequences of the **ProcM, LanT and LanOM** proteins from Prochlorococcus **MIT9313** were used as queries for the search of homologous proteins. \n",
    "3. Top hits were further inspected against the **NCBI Conserved Domain Database (CDD)** (13) to verify the presence of signature domains.\n",
    "4. The **tBLASTn** algorithm was used to search **procA and procA pseudogenes** using a **consensus sequence** derived from **29 prochlorosin leader peptide** sequences from Prochlorococcus **MIT9313**. \n",
    "5. BLAST hits with **>45% identity** and **E-val <0.001** were inspected against the CDD to confirm the presence of the **nif11-like leader peptide domain**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "![](http://oregonstate.edu/instruct/bb450/fall14/stryer7/2/table_02_02.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.uniprot.org/uniprot/Q7V735\n",
    "    http://tigrfams.jcvi.org/cgi-bin/HmmReportPage.cgi?acc=TIGR03798\n",
    "        https://www.ebi.ac.uk/training/online/course/interpro-functional-and-structural-analysis-protei/sequence-searching/searching-interpro-batc\n",
    "            http://www.ebi.ac.uk/interpro/sequencesearch/iprscan5-S20190707-131508-0462-76111813-p1m\n",
    "            https://www.ebi.ac.uk/Tools/services/rest/iprscan5/result/iprscan5-S20190707-131508-0462-76111813-p1m/json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_dpath = r'data\\detailed_Prochlorococcus_genome_annotations'\n",
    "genome_fnames = [ n for n in os.listdir(genomes_dpath) if n.endswith('.txt')]\n",
    "def _load(fname):\n",
    "    df = pd.read_csv(os.path.join(genomes_dpath, fname), sep='\\t')\n",
    "    df['genome'] = os.path.basename(os.path.splitext(fname)[0])\n",
    "    return df\n",
    "genome_df = pd.concat([_load(fname) for fname in genome_fnames])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>function</th>\n",
       "      <th>aliases</th>\n",
       "      <th>figfam</th>\n",
       "      <th>evidence_codes</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "      <th>genome</th>\n",
       "      <th>aa_length</th>\n",
       "      <th>gg_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AS9601</td>\n",
       "      <td>PAS9601_0001</td>\n",
       "      <td>fig|1218.83.peg.1</td>\n",
       "      <td>peg</td>\n",
       "      <td>AS9601_168_1325</td>\n",
       "      <td>168</td>\n",
       "      <td>1325</td>\n",
       "      <td>+</td>\n",
       "      <td>DNA polymerase III beta subunit (EC 2.7.7.7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG00066425</td>\n",
       "      <td>isu;DNA-replication isu;DNA_replication_cluster_1</td>\n",
       "      <td>atggaaattatttgtaatcaaaatgaattaaataatgctatacaac...</td>\n",
       "      <td>MEIICNQNELNNAIQLVSKAVASRPTHPILANILLTADEGTNKISV...</td>\n",
       "      <td>AS9601</td>\n",
       "      <td>385.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AS9601</td>\n",
       "      <td>PAS9601_0002</td>\n",
       "      <td>fig|1218.83.peg.2</td>\n",
       "      <td>peg</td>\n",
       "      <td>AS9601_1327_2034</td>\n",
       "      <td>1327</td>\n",
       "      <td>2034</td>\n",
       "      <td>+</td>\n",
       "      <td>RNA metabolism-related protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ttgaaattacctaaagaaattttattaagtgaattattaaattata...</td>\n",
       "      <td>MKLPKEILLSELLNYIVKGNMVLNYGNGENVWMHPPVHRILGWYSR...</td>\n",
       "      <td>AS9601</td>\n",
       "      <td>235.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AS9601</td>\n",
       "      <td>PAS9601_0003</td>\n",
       "      <td>fig|1218.83.peg.3</td>\n",
       "      <td>peg</td>\n",
       "      <td>AS9601_2038_4377</td>\n",
       "      <td>2038</td>\n",
       "      <td>4377</td>\n",
       "      <td>+</td>\n",
       "      <td>Phosphoribosylformylglycinamidine synthase, sy...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG01303876</td>\n",
       "      <td>icw(1);De_Novo_Purine_Biosynthesis</td>\n",
       "      <td>atgataaatcatgaaaataatgatctatttgatcttaatgaagcat...</td>\n",
       "      <td>MINHENNDLFDLNEALKVENLTLNDYEEICKRLKRKPNRTELGMFG...</td>\n",
       "      <td>AS9601</td>\n",
       "      <td>779.0</td>\n",
       "      <td>114.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AS9601</td>\n",
       "      <td>PAS9601_0004</td>\n",
       "      <td>fig|1218.83.peg.4</td>\n",
       "      <td>peg</td>\n",
       "      <td>AS9601_4425_5885</td>\n",
       "      <td>4425</td>\n",
       "      <td>5885</td>\n",
       "      <td>+</td>\n",
       "      <td>Amidophosphoribosyltransferase (EC 2.4.2.14)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG00000179</td>\n",
       "      <td>isu;YgfZ isu;De_Novo_Purine_Biosynthesis</td>\n",
       "      <td>atgtgcggaatagttggaatcgtttcttcgaatgatgtaaatcaac...</td>\n",
       "      <td>MCGIVGIVSSNDVNQQIYDSLLLLQHRGQDSTGIATMENTVFHIHK...</td>\n",
       "      <td>AS9601</td>\n",
       "      <td>486.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AS9601</td>\n",
       "      <td>PAS9601_0005</td>\n",
       "      <td>fig|1218.83.peg.5</td>\n",
       "      <td>peg</td>\n",
       "      <td>AS9601_8323_5882</td>\n",
       "      <td>8323</td>\n",
       "      <td>5882</td>\n",
       "      <td>-</td>\n",
       "      <td>DNA gyrase subunit A (EC 5.99.1.3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG00000080</td>\n",
       "      <td>idu(1);DNA_gyrase_subunits idu(1);DNA_topoisom...</td>\n",
       "      <td>atggataagaaaaatttcacttccatatcacttcaagaagaaatgc...</td>\n",
       "      <td>MDKKNFTSISLQEEMQRSYLEYAMSVIVGRALPDARDGLKPVQRRI...</td>\n",
       "      <td>AS9601</td>\n",
       "      <td>813.0</td>\n",
       "      <td>218.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  contig_id       gene_id         feature_id type          location  start  \\\n",
       "0    AS9601  PAS9601_0001  fig|1218.83.peg.1  peg   AS9601_168_1325    168   \n",
       "1    AS9601  PAS9601_0002  fig|1218.83.peg.2  peg  AS9601_1327_2034   1327   \n",
       "2    AS9601  PAS9601_0003  fig|1218.83.peg.3  peg  AS9601_2038_4377   2038   \n",
       "3    AS9601  PAS9601_0004  fig|1218.83.peg.4  peg  AS9601_4425_5885   4425   \n",
       "4    AS9601  PAS9601_0005  fig|1218.83.peg.5  peg  AS9601_8323_5882   8323   \n",
       "\n",
       "   stop strand                                           function  aliases  \\\n",
       "0  1325      +       DNA polymerase III beta subunit (EC 2.7.7.7)      NaN   \n",
       "1  2034      +                     RNA metabolism-related protein      NaN   \n",
       "2  4377      +  Phosphoribosylformylglycinamidine synthase, sy...      NaN   \n",
       "3  5885      +       Amidophosphoribosyltransferase (EC 2.4.2.14)      NaN   \n",
       "4  5882      -                 DNA gyrase subunit A (EC 5.99.1.3)      NaN   \n",
       "\n",
       "        figfam                                     evidence_codes  \\\n",
       "0  FIG00066425  isu;DNA-replication isu;DNA_replication_cluster_1   \n",
       "1          NaN                                                NaN   \n",
       "2  FIG01303876                 icw(1);De_Novo_Purine_Biosynthesis   \n",
       "3  FIG00000179           isu;YgfZ isu;De_Novo_Purine_Biosynthesis   \n",
       "4  FIG00000080  idu(1);DNA_gyrase_subunits idu(1);DNA_topoisom...   \n",
       "\n",
       "                                 nucleotide_sequence  \\\n",
       "0  atggaaattatttgtaatcaaaatgaattaaataatgctatacaac...   \n",
       "1  ttgaaattacctaaagaaattttattaagtgaattattaaattata...   \n",
       "2  atgataaatcatgaaaataatgatctatttgatcttaatgaagcat...   \n",
       "3  atgtgcggaatagttggaatcgtttcttcgaatgatgtaaatcaac...   \n",
       "4  atggataagaaaaatttcacttccatatcacttcaagaagaaatgc...   \n",
       "\n",
       "                                         aa_sequence  genome  aa_length  \\\n",
       "0  MEIICNQNELNNAIQLVSKAVASRPTHPILANILLTADEGTNKISV...  AS9601      385.0   \n",
       "1  MKLPKEILLSELLNYIVKGNMVLNYGNGENVWMHPPVHRILGWYSR...  AS9601      235.0   \n",
       "2  MINHENNDLFDLNEALKVENLTLNDYEEICKRLKRKPNRTELGMFG...  AS9601      779.0   \n",
       "3  MCGIVGIVSSNDVNQQIYDSLLLLQHRGQDSTGIATMENTVFHIHK...  AS9601      486.0   \n",
       "4  MDKKNFTSISLQEEMQRSYLEYAMSVIVGRALPDARDGLKPVQRRI...  AS9601      813.0   \n",
       "\n",
       "   gg_index  \n",
       "0      -1.0  \n",
       "1      -1.0  \n",
       "2     114.0  \n",
       "3      -1.0  \n",
       "4     218.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_df['aa_length'] = genome_df['aa_sequence'].str.len()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_df['gg_index'] = genome_df['aa_sequence'].str.find('GG')\n",
    "genome_df['ga_index'] = genome_df['aa_sequence'].str.find('GA')\n",
    "def min_index(gg_index, ga_index):\n",
    "    if gg_index == -1: \n",
    "        return ga_index\n",
    "    if ga_index == -1: \n",
    "        return gg_index\n",
    "    return min(ga_index, gg_index)\n",
    "genome_df['leader_end_index'] = min_index(genome_df['ga_index'], genome_df['gg_index'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_df['aa_sequence_post_gg'] = \n",
    "genome_df['SCT'] = genome_df['aa_sequence'].str.find('GG')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
