{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import sys, os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.special import comb\n",
    "from scipy import stats\n",
    "import scipy.cluster.hierarchy as hac\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('classic')\n",
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from collections import Counter\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "sns.set(rc={'figure.figsize':(15,8)})\n",
    "sns.set_context('poster')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "#import CCPA_lib as cp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context('poster')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Catalytic Promiscuity in the Biosynthesis of Cyclic Peptide in Planktonic Marine Cyanobacteria, PNAS2010\n",
    "Detection of procM and procA gene sequences in marine cyanobacterial genomes.\n",
    "1. **procM** gene sequences were detected in marine cyanobacterial genomes using the **pre-computed phylogenetic tree viewer in the microbesonline database** (http://www.microbesonline.org (3), which is computed using FastBLAST (4), see http://www.microbesonline.org/treebrowseHelp.html for additional details). \n",
    "2. **procA** sequences were detected in the genomes of Prochlorococcus **MIT9313** by **manually searching for short ORFs close to the procM gene** that encoded a **peptide** with a **leader sequence** followed by a **double-glycine** protease recognition site and a **short** core peptide rich in **Cys, Ser and Thr** residues (Fig. 1C). \n",
    "3. Once several such putative lantipeptides were discovered, we used **BLASTp** in both **microbesonline** and the **NCBI nr** database (with a cutoff **e-value of 10**) to detect other, similar genes in cyanobacterial genomes\n",
    "4.  **manually** checked the resulting peptide sequences to verify that they agreed with the definition of procAs, as described above. \n",
    "5. Figure 1 and Figure S1 were produced using IMG (http://img.jgi.doe.gov/cgi-bin/pub/main.cgi)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evolutionary radiation of lanthipeptides in marine cyanobacteria, PNAPLUS 2017\n",
    "1. Annotation of **prochlorosin** biosynthesis genes was performed manually using **BLAST** (12). \n",
    "2. The sequences of the **ProcM, LanT and LanOM** proteins from Prochlorococcus **MIT9313** were used as queries for the search of homologous proteins. \n",
    "3. Top hits were further inspected against the **NCBI Conserved Domain Database (CDD)** (13) to verify the presence of signature domains.\n",
    "4. The **tBLASTn** algorithm was used to search **procA and procA pseudogenes** using a **consensus sequence** derived from **29 prochlorosin leader peptide** sequences from Prochlorococcus **MIT9313**. \n",
    "5. BLAST hits with **>45% identity** and **E-val <0.001** were inspected against the CDD to confirm the presence of the **nif11-like leader peptide domain**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "![](http://oregonstate.edu/instruct/bb450/fall14/stryer7/2/table_02_02.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.uniprot.org/uniprot/Q7V735\n",
    "\n",
    "http://tigrfams.jcvi.org/cgi-bin/HmmReportPage.cgi?acc=TIGR03798\n",
    "\n",
    "https://www.ebi.ac.uk/training/online/course/interpro-functional-and-structural-analysis-protei/sequence-searching/searching-interpro-batc\n",
    "\n",
    "http://www.ebi.ac.uk/interpro/sequencesearch/iprscan5-S20190707-131508-0462-76111813-p1m\n",
    "\n",
    "https://www.ebi.ac.uk/Tools/services/rest/iprscan5/result/iprscan5-S20190707-131508-0462-76111813-p1m/json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_split_index(aa_sequence):\n",
    "    gg_index = aa_sequence.find('GG')\n",
    "    return gg_index\n",
    "    ga_index = aa_sequence.find('GA')\n",
    "    split_index = -1\n",
    "    if gg_index == -1:\n",
    "        split_index = ga_index\n",
    "    elif ga_index == -1:\n",
    "        split_index = gg_index\n",
    "    else:\n",
    "        split_index = min(gg_index, ga_index)\n",
    "    return split_index\n",
    "    \n",
    "def analyze_procm_data(row):\n",
    "    aa_sequence = str(row['aa_sequence'])\n",
    "    function = str(row['function'])\n",
    "    aa_length = len(aa_sequence)\n",
    "    split_index = find_split_index(aa_sequence)\n",
    "    core_sequence = None\n",
    "    core_length = None\n",
    "    count_dict = dict()\n",
    "    precent_dict = dict()\n",
    "    is_lanm = False\n",
    "    if function == 'Lanthionine biosynthesis protein LanM':\n",
    "        is_lanm = True\n",
    "\n",
    "    if split_index != -1:\n",
    "        core_sequence = aa_sequence[split_index+2:]\n",
    "        core_length = len(core_sequence)\n",
    "        if core_length > 0:\n",
    "            count = Counter(core_sequence)\n",
    "            count_dict = {f'core_count_{k}' : v for k,v in count.items()}\n",
    "            precent_dict = {f'core_precent_{k}' : v/core_length for k,v in count.items()}\n",
    "    return row.append(pd.Series({\n",
    "        'aa_length' : aa_length,\n",
    "        'split_index' : split_index,\n",
    "        'core_sequence' : core_sequence,\n",
    "        'core_length' : core_length,\n",
    "        'is_lanm' : is_lanm,\n",
    "        **count_dict, \n",
    "        **precent_dict, \n",
    "    }))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "def run_diamond_to_create_db():\n",
    "    dbname='pro_faa'\n",
    "    cwd = os.getcwd()\n",
    "    in_fasta_dpath = os.path.join(cwd, 'data', 'all_Prochlorococcus_ORFs_faa')\n",
    "    \n",
    "    diamond_exe = os.path.join(cwd, 'bin','diamond.exe')\n",
    "    out_dpath = os.path.join(cwd,'diamond_tmp')\n",
    "    diamond_params = [#'TYPE', os.path.join(in_fasta_dpath, '*'), \n",
    "                      #'|', \n",
    "                      diamond_exe, 'makedb', \n",
    "                      '--in', os.path.join(in_fasta_dpath, 'all_faa.faa'), \n",
    "                      '-d', os.path.join(out_dpath, dbname)\n",
    "    ]\n",
    "    try:\n",
    "        os.makedirs(out_dpath, exist_ok=True)\n",
    "        retcode = subprocess.check_call(' '.join(diamond_params), shell=True)\n",
    "        if retcode < 0:\n",
    "            print(\"Child was terminated by signal\", -retcode)\n",
    "    except OSError as e:\n",
    "        print(\"Execution failed:\", e)\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(\"Execution failed:\", e)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "def _write_fasta(fname, sequence):\n",
    "    with open(fname, 'w') as f:\n",
    "        f.write(f'> tmp\\n{sequence}\\n')\n",
    "        \n",
    "def _read_diamond_results(fname):\n",
    "    diamond_columns = [ 'gene_id', 'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart',\n",
    "                   'send', 'evalue', 'bitscore']\n",
    "    dout_df = pd.read_csv(fname, sep='\\t', \n",
    "                     names=diamond_columns)\n",
    "    return dout_df.reset_index(drop=True)\n",
    "\n",
    "def run_diamond_to_align_sequence(aa_sequence):\n",
    "    dbname='pro_faa'\n",
    "    cwd = os.getcwd()\n",
    "    in_fasta_dpath = os.path.join(cwd, 'data', 'all_Prochlorococcus_ORFs_faa')\n",
    "    \n",
    "    diamond_exe = os.path.join(cwd, 'bin','diamond.exe')\n",
    "    out_dpath = os.path.join(cwd,'diamond_tmp')\n",
    "    out_fpath = os.path.join(out_dpath, 'diamond_out.txt')\n",
    "    tmp_fpath = os.path.join(out_dpath, 'tmp.faa')\n",
    "\n",
    "    os.makedirs(out_dpath, exist_ok=True)\n",
    "    _write_fasta(tmp_fpath, aa_sequence)\n",
    "    diamond_params = [#'TYPE', os.path.join(in_fasta_dpath, '*'), \n",
    "                      #'|', \n",
    "                      diamond_exe, 'blastp', \n",
    "                      '-q', tmp_fpath, \n",
    "                      '-d', os.path.join(out_dpath, dbname),\n",
    "                      '-o', out_fpath,\n",
    "    ]\n",
    "    try:\n",
    "        subprocess.check_call(' '.join(diamond_params), shell=True)\n",
    "        return _read_diamond_results(out_fpath)\n",
    "    except OSError as e:\n",
    "        print(\"Execution failed:\", e)\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(\"Execution failed:\", e)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_dpath = r'data\\detailed_Prochlorococcus_genome_annotations'\n",
    "genome_fnames = [ n for n in os.listdir(genomes_dpath) if n.endswith('.txt')]\n",
    "def _load(fname):\n",
    "    df = pd.read_csv(os.path.join(genomes_dpath, fname), sep='\\t')\n",
    "    df['genome'] = os.path.basename(os.path.splitext(fname)[0])\n",
    "    return df\n",
    "genome_df = pd.concat([_load(fname) for fname in genome_fnames])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_df  = genome_df.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>type</th>\n",
       "      <th>location</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>function</th>\n",
       "      <th>aliases</th>\n",
       "      <th>figfam</th>\n",
       "      <th>evidence_codes</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>aa_sequence</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AS9601</td>\n",
       "      <td>PAS9601_0001</td>\n",
       "      <td>fig|1218.83.peg.1</td>\n",
       "      <td>peg</td>\n",
       "      <td>AS9601_168_1325</td>\n",
       "      <td>168</td>\n",
       "      <td>1325</td>\n",
       "      <td>+</td>\n",
       "      <td>DNA polymerase III beta subunit (EC 2.7.7.7)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG00066425</td>\n",
       "      <td>isu;DNA-replication isu;DNA_replication_cluster_1</td>\n",
       "      <td>atggaaattatttgtaatcaaaatgaattaaataatgctatacaac...</td>\n",
       "      <td>MEIICNQNELNNAIQLVSKAVASRPTHPILANILLTADEGTNKISV...</td>\n",
       "      <td>AS9601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AS9601</td>\n",
       "      <td>PAS9601_0002</td>\n",
       "      <td>fig|1218.83.peg.2</td>\n",
       "      <td>peg</td>\n",
       "      <td>AS9601_1327_2034</td>\n",
       "      <td>1327</td>\n",
       "      <td>2034</td>\n",
       "      <td>+</td>\n",
       "      <td>RNA metabolism-related protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ttgaaattacctaaagaaattttattaagtgaattattaaattata...</td>\n",
       "      <td>MKLPKEILLSELLNYIVKGNMVLNYGNGENVWMHPPVHRILGWYSR...</td>\n",
       "      <td>AS9601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AS9601</td>\n",
       "      <td>PAS9601_0003</td>\n",
       "      <td>fig|1218.83.peg.3</td>\n",
       "      <td>peg</td>\n",
       "      <td>AS9601_2038_4377</td>\n",
       "      <td>2038</td>\n",
       "      <td>4377</td>\n",
       "      <td>+</td>\n",
       "      <td>Phosphoribosylformylglycinamidine synthase, sy...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG01303876</td>\n",
       "      <td>icw(1);De_Novo_Purine_Biosynthesis</td>\n",
       "      <td>atgataaatcatgaaaataatgatctatttgatcttaatgaagcat...</td>\n",
       "      <td>MINHENNDLFDLNEALKVENLTLNDYEEICKRLKRKPNRTELGMFG...</td>\n",
       "      <td>AS9601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AS9601</td>\n",
       "      <td>PAS9601_0004</td>\n",
       "      <td>fig|1218.83.peg.4</td>\n",
       "      <td>peg</td>\n",
       "      <td>AS9601_4425_5885</td>\n",
       "      <td>4425</td>\n",
       "      <td>5885</td>\n",
       "      <td>+</td>\n",
       "      <td>Amidophosphoribosyltransferase (EC 2.4.2.14)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG00000179</td>\n",
       "      <td>isu;YgfZ isu;De_Novo_Purine_Biosynthesis</td>\n",
       "      <td>atgtgcggaatagttggaatcgtttcttcgaatgatgtaaatcaac...</td>\n",
       "      <td>MCGIVGIVSSNDVNQQIYDSLLLLQHRGQDSTGIATMENTVFHIHK...</td>\n",
       "      <td>AS9601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AS9601</td>\n",
       "      <td>PAS9601_0005</td>\n",
       "      <td>fig|1218.83.peg.5</td>\n",
       "      <td>peg</td>\n",
       "      <td>AS9601_8323_5882</td>\n",
       "      <td>8323</td>\n",
       "      <td>5882</td>\n",
       "      <td>-</td>\n",
       "      <td>DNA gyrase subunit A (EC 5.99.1.3)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FIG00000080</td>\n",
       "      <td>idu(1);DNA_gyrase_subunits idu(1);DNA_topoisom...</td>\n",
       "      <td>atggataagaaaaatttcacttccatatcacttcaagaagaaatgc...</td>\n",
       "      <td>MDKKNFTSISLQEEMQRSYLEYAMSVIVGRALPDARDGLKPVQRRI...</td>\n",
       "      <td>AS9601</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  contig_id       gene_id         feature_id type          location  start  \\\n",
       "0    AS9601  PAS9601_0001  fig|1218.83.peg.1  peg   AS9601_168_1325    168   \n",
       "1    AS9601  PAS9601_0002  fig|1218.83.peg.2  peg  AS9601_1327_2034   1327   \n",
       "2    AS9601  PAS9601_0003  fig|1218.83.peg.3  peg  AS9601_2038_4377   2038   \n",
       "3    AS9601  PAS9601_0004  fig|1218.83.peg.4  peg  AS9601_4425_5885   4425   \n",
       "4    AS9601  PAS9601_0005  fig|1218.83.peg.5  peg  AS9601_8323_5882   8323   \n",
       "\n",
       "   stop strand                                           function  aliases  \\\n",
       "0  1325      +       DNA polymerase III beta subunit (EC 2.7.7.7)      NaN   \n",
       "1  2034      +                     RNA metabolism-related protein      NaN   \n",
       "2  4377      +  Phosphoribosylformylglycinamidine synthase, sy...      NaN   \n",
       "3  5885      +       Amidophosphoribosyltransferase (EC 2.4.2.14)      NaN   \n",
       "4  5882      -                 DNA gyrase subunit A (EC 5.99.1.3)      NaN   \n",
       "\n",
       "        figfam                                     evidence_codes  \\\n",
       "0  FIG00066425  isu;DNA-replication isu;DNA_replication_cluster_1   \n",
       "1          NaN                                                NaN   \n",
       "2  FIG01303876                 icw(1);De_Novo_Purine_Biosynthesis   \n",
       "3  FIG00000179           isu;YgfZ isu;De_Novo_Purine_Biosynthesis   \n",
       "4  FIG00000080  idu(1);DNA_gyrase_subunits idu(1);DNA_topoisom...   \n",
       "\n",
       "                                 nucleotide_sequence  \\\n",
       "0  atggaaattatttgtaatcaaaatgaattaaataatgctatacaac...   \n",
       "1  ttgaaattacctaaagaaattttattaagtgaattattaaattata...   \n",
       "2  atgataaatcatgaaaataatgatctatttgatcttaatgaagcat...   \n",
       "3  atgtgcggaatagttggaatcgtttcttcgaatgatgtaaatcaac...   \n",
       "4  atggataagaaaaatttcacttccatatcacttcaagaagaaatgc...   \n",
       "\n",
       "                                         aa_sequence  genome  \n",
       "0  MEIICNQNELNNAIQLVSKAVASRPTHPILANILLTADEGTNKISV...  AS9601  \n",
       "1  MKLPKEILLSELLNYIVKGNMVLNYGNGENVWMHPPVHRILGWYSR...  AS9601  \n",
       "2  MINHENNDLFDLNEALKVENLTLNDYEEICKRLKRKPNRTELGMFG...  AS9601  \n",
       "3  MCGIVGIVSSNDVNQQIYDSLLLLQHRGQDSTGIATMENTVFHIHK...  AS9601  \n",
       "4  MDKKNFTSISLQEEMQRSYLEYAMSVIVGRALPDARDGLKPVQRRI...  AS9601  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MSEEQLKAFIAKVQADTSLQEQLKAEGADVVAIAKAAGFSITTEDLEKEHRQTLSDDDLEGVAGGFFCVQGTANRFTINVC'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_df.iloc[59275].aa_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = genome_df.apply(analyze_procm_data, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_candidate(r, \n",
    "                 max_aa_sequence_length = 600,\n",
    "                 min_core_length = 5, \n",
    "                 max_core_length = 200, \n",
    "                 min_split_index = 5,\n",
    "                 min_prec_CST = 0.1\n",
    "                ):\n",
    "    row = r.fillna(0)\n",
    "    if ((row.split_index > min_split_index) and \n",
    "        (row.aa_length < max_aa_sequence_length) and \n",
    "        (row.core_length < max_core_length) and\n",
    "        (row.core_length > min_core_length) and\n",
    "        (row.core_precent_C + row.core_precent_T + row.core_precent_S > min_prec_CST)\n",
    "       ):\n",
    "        return True\n",
    "    return False\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['is_candidate'] = df.apply(is_candidate, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genome\n",
       "AS9601     384.0\n",
       "EQPAC1     391.0\n",
       "GP2        368.0\n",
       "HNLC1      353.0\n",
       "HNLC2      347.0\n",
       "LG         429.0\n",
       "MED4       392.0\n",
       "MIT0601    432.0\n",
       "MIT0602    433.0\n",
       "MIT0603    431.0\n",
       "MIT0604    417.0\n",
       "MIT0701    636.0\n",
       "MIT0702    623.0\n",
       "MIT0703    624.0\n",
       "MIT0801    436.0\n",
       "MIT9107    386.0\n",
       "MIT9116    387.0\n",
       "MIT9123    386.0\n",
       "MIT9201    389.0\n",
       "MIT9202    381.0\n",
       "MIT9211    419.0\n",
       "MIT9215    392.0\n",
       "MIT9301    394.0\n",
       "MIT9302    408.0\n",
       "MIT9303    639.0\n",
       "MIT9311    389.0\n",
       "MIT9312    392.0\n",
       "MIT9313    637.0\n",
       "MIT9314    393.0\n",
       "MIT9321    386.0\n",
       "MIT9322    394.0\n",
       "MIT9401    389.0\n",
       "MIT9515    374.0\n",
       "NATL1A     437.0\n",
       "NATL2A     406.0\n",
       "PAC1       439.0\n",
       "SB         381.0\n",
       "SS120      423.0\n",
       "SS2        429.0\n",
       "SS35       427.0\n",
       "SS51       421.0\n",
       "SS52       429.0\n",
       "UH18301    392.0\n",
       "Name: is_candidate, dtype: float64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby('genome').is_candidate.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected = df.loc[df.function.str.contains('lanm', case=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa_length</th>\n",
       "      <th>aa_sequence</th>\n",
       "      <th>aliases</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>core_count_A</th>\n",
       "      <th>core_count_C</th>\n",
       "      <th>core_count_D</th>\n",
       "      <th>core_count_E</th>\n",
       "      <th>core_count_F</th>\n",
       "      <th>core_count_G</th>\n",
       "      <th>...</th>\n",
       "      <th>is_lanm</th>\n",
       "      <th>location</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>split_index</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>type</th>\n",
       "      <th>dist_after_cumsum</th>\n",
       "      <th>is_candidate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23489</th>\n",
       "      <td>1082</td>\n",
       "      <td>MGAAEPLALQSSWKSLWLEAVAADEPHKFKRRLEWDGLTESEFHAA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>contig00017</td>\n",
       "      <td>91.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>contig00017_4336_7584</td>\n",
       "      <td>atgggagctgctgagcctttggcacttcaatcatcttggaaatccc...</td>\n",
       "      <td>255</td>\n",
       "      <td>4336</td>\n",
       "      <td>7584</td>\n",
       "      <td>+</td>\n",
       "      <td>peg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26640</th>\n",
       "      <td>1082</td>\n",
       "      <td>MGAAEPLALQSSWKSLWLEAVAADEPHKFKRRLEWDGLTESEFHAA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>contig00012</td>\n",
       "      <td>91.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>contig00012_65496_62248</td>\n",
       "      <td>atgggagctgctgagcctttggcacttcaatcatcttggaaatccc...</td>\n",
       "      <td>255</td>\n",
       "      <td>65496</td>\n",
       "      <td>62248</td>\n",
       "      <td>-</td>\n",
       "      <td>peg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30660</th>\n",
       "      <td>1082</td>\n",
       "      <td>MGAAEPLALQSSWKSLWLEAVAADEPHKFKRRLEWDGLTESEFHAA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>contig00032</td>\n",
       "      <td>91.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>contig00032_21482_18234</td>\n",
       "      <td>atgggagctgctgagcctttggcacttcaatcatcttggaaatccc...</td>\n",
       "      <td>255</td>\n",
       "      <td>21482</td>\n",
       "      <td>18234</td>\n",
       "      <td>-</td>\n",
       "      <td>peg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53940</th>\n",
       "      <td>1074</td>\n",
       "      <td>MGAAEPLTSPTSWKTSWLAAIAPDEPHKFDRRLEWDELSEENFFAA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MIT9303</td>\n",
       "      <td>100.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT9303_1865514_1868738</td>\n",
       "      <td>atgggagctgctgagcctttgacaagtccaacatcttggaaaacca...</td>\n",
       "      <td>249</td>\n",
       "      <td>1865514</td>\n",
       "      <td>1868738</td>\n",
       "      <td>+</td>\n",
       "      <td>peg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59279</th>\n",
       "      <td>1074</td>\n",
       "      <td>MGAAEPLESPSSWKTSWLAAIAPDEPHKFDRRLEWDELSEENFFAA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MIT9313</td>\n",
       "      <td>93.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>MIT9313_283359_280135</td>\n",
       "      <td>atgggagctgccgagcctttggaaagtccatcatcttggaaaacca...</td>\n",
       "      <td>249</td>\n",
       "      <td>283359</td>\n",
       "      <td>280135</td>\n",
       "      <td>-</td>\n",
       "      <td>peg</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       aa_length                                        aa_sequence  aliases  \\\n",
       "23489       1082  MGAAEPLALQSSWKSLWLEAVAADEPHKFKRRLEWDGLTESEFHAA...      NaN   \n",
       "26640       1082  MGAAEPLALQSSWKSLWLEAVAADEPHKFKRRLEWDGLTESEFHAA...      NaN   \n",
       "30660       1082  MGAAEPLALQSSWKSLWLEAVAADEPHKFKRRLEWDGLTESEFHAA...      NaN   \n",
       "53940       1074  MGAAEPLTSPTSWKTSWLAAIAPDEPHKFDRRLEWDELSEENFFAA...      NaN   \n",
       "59279       1074  MGAAEPLESPSSWKTSWLAAIAPDEPHKFDRRLEWDELSEENFFAA...      NaN   \n",
       "\n",
       "         contig_id  core_count_A  core_count_C  core_count_D  core_count_E  \\\n",
       "23489  contig00017          91.0          19.0          50.0          45.0   \n",
       "26640  contig00012          91.0          19.0          50.0          45.0   \n",
       "30660  contig00032          91.0          19.0          50.0          45.0   \n",
       "53940      MIT9303         100.0          15.0          47.0          43.0   \n",
       "59279      MIT9313          93.0          15.0          47.0          45.0   \n",
       "\n",
       "       core_count_F  core_count_G  ...  is_lanm                 location  \\\n",
       "23489          27.0          67.0  ...     True    contig00017_4336_7584   \n",
       "26640          27.0          67.0  ...     True  contig00012_65496_62248   \n",
       "30660          27.0          67.0  ...     True  contig00032_21482_18234   \n",
       "53940          29.0          66.0  ...     True  MIT9303_1865514_1868738   \n",
       "59279          29.0          68.0  ...     True    MIT9313_283359_280135   \n",
       "\n",
       "                                     nucleotide_sequence  split_index  \\\n",
       "23489  atgggagctgctgagcctttggcacttcaatcatcttggaaatccc...          255   \n",
       "26640  atgggagctgctgagcctttggcacttcaatcatcttggaaatccc...          255   \n",
       "30660  atgggagctgctgagcctttggcacttcaatcatcttggaaatccc...          255   \n",
       "53940  atgggagctgctgagcctttgacaagtccaacatcttggaaaacca...          249   \n",
       "59279  atgggagctgccgagcctttggaaagtccatcatcttggaaaacca...          249   \n",
       "\n",
       "         start     stop  strand  type  dist_after_cumsum  is_candidate  \n",
       "23489     4336     7584       +   peg               True         False  \n",
       "26640    65496    62248       -   peg               True         False  \n",
       "30660    21482    18234       -   peg               True         False  \n",
       "53940  1865514  1868738       +   peg               True         False  \n",
       "59279   283359   280135       -   peg               True         False  \n",
       "\n",
       "[5 rows x 66 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected = df.loc[df.gene_id == 'PMIT9313_0302']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "59279    PMIT9313_0302\n",
       "Name: gene_id, dtype: object"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected.gene_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>59261</th>\n",
       "      <td>PMIT9313_0284</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59262</th>\n",
       "      <td>PMIT9313_0285</td>\n",
       "      <td>possible Indole-3-glycerol phosphate synthase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59264</th>\n",
       "      <td>PMIT9313_0287</td>\n",
       "      <td>possible Indole-3-glycerol phosphate synthase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59267</th>\n",
       "      <td>PMIT9313_0290</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59268</th>\n",
       "      <td>PMIT9313_0291</td>\n",
       "      <td>possible [2Fe-2S] binding domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59269</th>\n",
       "      <td>PMIT9313_0292</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59271</th>\n",
       "      <td>PMIT9313_0294</td>\n",
       "      <td>possible (M20568) ORF 11 [Azotobacter vinelandii]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59273</th>\n",
       "      <td>PMIT9313_0296</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59274</th>\n",
       "      <td>PMIT9313_0297</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59275</th>\n",
       "      <td>PMIT9313_0298</td>\n",
       "      <td>possible (M20568) ORF 11 [Azotobacter vinelandii]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59281</th>\n",
       "      <td>PMIT9313_0304</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59285</th>\n",
       "      <td>PMIT9313_0308</td>\n",
       "      <td>NifU-like domain protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59293</th>\n",
       "      <td>PMIT9313_0316</td>\n",
       "      <td>FIG00942299: hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59296</th>\n",
       "      <td>PMIT9313_0319</td>\n",
       "      <td>FIG00940919: hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59298</th>\n",
       "      <td>PMIT9313_0321</td>\n",
       "      <td>possible pilin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id                                           function\n",
       "59261  PMIT9313_0284                               hypothetical protein\n",
       "59262  PMIT9313_0285      possible Indole-3-glycerol phosphate synthase\n",
       "59264  PMIT9313_0287      possible Indole-3-glycerol phosphate synthase\n",
       "59267  PMIT9313_0290                               hypothetical protein\n",
       "59268  PMIT9313_0291                   possible [2Fe-2S] binding domain\n",
       "59269  PMIT9313_0292                               hypothetical protein\n",
       "59271  PMIT9313_0294  possible (M20568) ORF 11 [Azotobacter vinelandii]\n",
       "59273  PMIT9313_0296                               hypothetical protein\n",
       "59274  PMIT9313_0297                               hypothetical protein\n",
       "59275  PMIT9313_0298  possible (M20568) ORF 11 [Azotobacter vinelandii]\n",
       "59281  PMIT9313_0304                               hypothetical protein\n",
       "59285  PMIT9313_0308                           NifU-like domain protein\n",
       "59293  PMIT9313_0316                  FIG00942299: hypothetical protein\n",
       "59296  PMIT9313_0319                  FIG00940919: hypothetical protein\n",
       "59298  PMIT9313_0321                                     possible pilin"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selection_win = 20\n",
    "idx = selected.index[0]\n",
    "idx\n",
    "vib = df.iloc[idx - selection_win: idx + selection_win + 1 ]\n",
    "vib[vib.is_candidate][['gene_id', 'function']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "277780 .. 278052 (-) \n",
    "276160 .. 276447 (-) \n",
    "275220 .. 275483 (-) \n",
    "274751 .. 274996 (-) \n",
    "274290 .. 274553 (-) \n",
    "273356 .. 273631 (-) \n",
    "272419 .. 272715 (-) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "row = row.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['aa_length', 'aa_sequence', 'aliases', 'contig_id', 'core_count_A',\n",
       "       'core_count_C', 'core_count_D', 'core_count_E', 'core_count_F',\n",
       "       'core_count_G', 'core_count_H', 'core_count_I', 'core_count_K',\n",
       "       'core_count_L', 'core_count_M', 'core_count_N', 'core_count_P',\n",
       "       'core_count_Q', 'core_count_R', 'core_count_S', 'core_count_T',\n",
       "       'core_count_V', 'core_count_W', 'core_count_X', 'core_count_Y',\n",
       "       'core_count_x', 'core_length', 'core_precent_A', 'core_precent_C',\n",
       "       'core_precent_D', 'core_precent_E', 'core_precent_F', 'core_precent_G',\n",
       "       'core_precent_H', 'core_precent_I', 'core_precent_K', 'core_precent_L',\n",
       "       'core_precent_M', 'core_precent_N', 'core_precent_P', 'core_precent_Q',\n",
       "       'core_precent_R', 'core_precent_S', 'core_precent_T', 'core_precent_V',\n",
       "       'core_precent_W', 'core_precent_X', 'core_precent_Y', 'core_precent_x',\n",
       "       'core_sequence', 'evidence_codes', 'feature_id', 'figfam', 'function',\n",
       "       'gene_id', 'genome', 'is_lanm', 'location', 'nucleotide_sequence',\n",
       "       'split_index', 'start', 'stop', 'strand', 'type', 'dist_after_cumsum',\n",
       "       'is_candidate'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_diamond_to_create_db('pro_faa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_diamond_to_align_sequence('MSEEQLKAFIAKVQADTSLQEQLKAEGADVVAIAKAAGFSITTEDLEKEHRQTLSDDDLEGVAGGFFCVQGTANRFTINVC')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "diamond_columns = [ 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart',\n",
    "                   'send', 'evalue', 'bitscore']\n",
    "dout_df = pd.read_csv(r'C:\\Users\\wosnat\\Documents\\GitHub\\ripp\\diamond_tmp\\diamond_out.txt', sep='\\t', \n",
    "                     names=diamond_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMIT9313_0298</td>\n",
       "      <td>100.0</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>81</td>\n",
       "      <td>1</td>\n",
       "      <td>81</td>\n",
       "      <td>3.100000e-39</td>\n",
       "      <td>158.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMIT9313_2802</td>\n",
       "      <td>92.6</td>\n",
       "      <td>68</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>67</td>\n",
       "      <td>6.800000e-26</td>\n",
       "      <td>114.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMIT9313_1204</td>\n",
       "      <td>95.4</td>\n",
       "      <td>65</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>65</td>\n",
       "      <td>1</td>\n",
       "      <td>64</td>\n",
       "      <td>2.600000e-25</td>\n",
       "      <td>112.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMIT9313_2803</td>\n",
       "      <td>87.0</td>\n",
       "      <td>69</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>69</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>7.500000e-25</td>\n",
       "      <td>110.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMIT9313_0287</td>\n",
       "      <td>89.4</td>\n",
       "      <td>66</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>65</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>1.700000e-24</td>\n",
       "      <td>109.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sseqid  pident  length  mismatch  gapopen  qstart  qend  sstart  \\\n",
       "0  PMIT9313_0298   100.0      81         0        0       1    81       1   \n",
       "1  PMIT9313_2802    92.6      68         4        1       1    68       1   \n",
       "2  PMIT9313_1204    95.4      65         2        1       1    65       1   \n",
       "3  PMIT9313_2803    87.0      69         8        1       1    69       1   \n",
       "4  PMIT9313_0287    89.4      66         6        1       1    65       1   \n",
       "\n",
       "   send        evalue  bitscore  \n",
       "0    81  3.100000e-39     158.7  \n",
       "1    67  6.800000e-26     114.4  \n",
       "2    64  2.600000e-25     112.5  \n",
       "3    68  7.500000e-25     110.9  \n",
       "4    66  1.700000e-24     109.8  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dout_df.reset_index(drop=True).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>59275</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aa_length</th>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aa_sequence</th>\n",
       "      <td>MSEEQLKAFIAKVQADTSLQEQLKAEGADVVAIAKAAGFSITTEDL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aliases</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>contig_id</th>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_A</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_C</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_D</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_E</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_F</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_G</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_H</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_I</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_K</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_L</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_M</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_N</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_P</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_Q</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_R</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_S</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_T</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_V</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_W</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_X</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_Y</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_count_x</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_length</th>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_A</th>\n",
       "      <td>0.113208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_C</th>\n",
       "      <td>0.0377358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_D</th>\n",
       "      <td>0.0943396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_I</th>\n",
       "      <td>0.0566038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_K</th>\n",
       "      <td>0.0377358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_L</th>\n",
       "      <td>0.0566038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_M</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_N</th>\n",
       "      <td>0.0377358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_P</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_Q</th>\n",
       "      <td>0.0377358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_R</th>\n",
       "      <td>0.0377358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_S</th>\n",
       "      <td>0.0377358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_T</th>\n",
       "      <td>0.0943396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_V</th>\n",
       "      <td>0.0943396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_W</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_X</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_Y</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_precent_x</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>core_sequence</th>\n",
       "      <td>DVVAIAKAAGFSITTEDLEKEHRQTLSDDDLEGVAGGFFCVQGTAN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>evidence_codes</th>\n",
       "      <td>ff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_id</th>\n",
       "      <td>fig|1218.91.peg.298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>figfam</th>\n",
       "      <td>FIG00940714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>function</th>\n",
       "      <td>possible (M20568) ORF 11 [Azotobacter vinelandii]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <td>PMIT9313_0298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome</th>\n",
       "      <td>MIT9313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_lanm</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>location</th>\n",
       "      <td>MIT9313_278025_277780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <td>atgtcagaagaacaactcaaggcattcattgccaaagttcaagccg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split_index</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start</th>\n",
       "      <td>278025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop</th>\n",
       "      <td>277780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strand</th>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>type</th>\n",
       "      <td>peg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                 59275\n",
       "aa_length                                                           81\n",
       "aa_sequence          MSEEQLKAFIAKVQADTSLQEQLKAEGADVVAIAKAAGFSITTEDL...\n",
       "aliases                                                            NaN\n",
       "contig_id                                                      MIT9313\n",
       "core_count_A                                                         6\n",
       "core_count_C                                                         2\n",
       "core_count_D                                                         5\n",
       "core_count_E                                                         4\n",
       "core_count_F                                                         4\n",
       "core_count_G                                                         5\n",
       "core_count_H                                                         1\n",
       "core_count_I                                                         3\n",
       "core_count_K                                                         2\n",
       "core_count_L                                                         3\n",
       "core_count_M                                                       NaN\n",
       "core_count_N                                                         2\n",
       "core_count_P                                                       NaN\n",
       "core_count_Q                                                         2\n",
       "core_count_R                                                         2\n",
       "core_count_S                                                         2\n",
       "core_count_T                                                         5\n",
       "core_count_V                                                         5\n",
       "core_count_W                                                       NaN\n",
       "core_count_X                                                       NaN\n",
       "core_count_Y                                                       NaN\n",
       "core_count_x                                                       NaN\n",
       "core_length                                                         53\n",
       "core_precent_A                                                0.113208\n",
       "core_precent_C                                               0.0377358\n",
       "core_precent_D                                               0.0943396\n",
       "...                                                                ...\n",
       "core_precent_I                                               0.0566038\n",
       "core_precent_K                                               0.0377358\n",
       "core_precent_L                                               0.0566038\n",
       "core_precent_M                                                     NaN\n",
       "core_precent_N                                               0.0377358\n",
       "core_precent_P                                                     NaN\n",
       "core_precent_Q                                               0.0377358\n",
       "core_precent_R                                               0.0377358\n",
       "core_precent_S                                               0.0377358\n",
       "core_precent_T                                               0.0943396\n",
       "core_precent_V                                               0.0943396\n",
       "core_precent_W                                                     NaN\n",
       "core_precent_X                                                     NaN\n",
       "core_precent_Y                                                     NaN\n",
       "core_precent_x                                                     NaN\n",
       "core_sequence        DVVAIAKAAGFSITTEDLEKEHRQTLSDDDLEGVAGGFFCVQGTAN...\n",
       "evidence_codes                                                      ff\n",
       "feature_id                                         fig|1218.91.peg.298\n",
       "figfam                                                     FIG00940714\n",
       "function             possible (M20568) ORF 11 [Azotobacter vinelandii]\n",
       "gene_id                                                  PMIT9313_0298\n",
       "genome                                                         MIT9313\n",
       "is_lanm                                                          False\n",
       "location                                         MIT9313_278025_277780\n",
       "nucleotide_sequence  atgtcagaagaacaactcaaggcattcattgccaaagttcaagccg...\n",
       "split_index                                                         26\n",
       "start                                                           278025\n",
       "stop                                                            277780\n",
       "strand                                                               -\n",
       "type                                                               peg\n",
       "\n",
       "[64 rows x 1 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.gene_id == 'PMIT9313_0298'].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(25, 11)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dout_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NaN    PMIT9313_0298\n",
       "NaN    PMIT9313_2802\n",
       "NaN    PMIT9313_1204\n",
       "NaN    PMIT9313_2803\n",
       "NaN    PMIT9313_0287\n",
       "NaN    PMIT9313_2818\n",
       "NaN    PMIT9313_2800\n",
       "NaN    PMIT9313_2801\n",
       "NaN    PMIT9313_1203\n",
       "NaN    PMIT9313_1047\n",
       "NaN    PMIT9303_0548\n",
       "NaN    PMIT9303_3044\n",
       "NaN    PMIT9313_1034\n",
       "NaN    PMIT9313_1037\n",
       "NaN    PMIT9313_2798\n",
       "NaN    PMIT9313_0291\n",
       "NaN    PMIT9313_2821\n",
       "NaN    PMIT9303_2965\n",
       "NaN    PMIT9303_1503\n",
       "NaN    PMIT9313_1222\n",
       "NaN    PMIT9313_0290\n",
       "NaN    PMIT9313_2816\n",
       "NaN    PMIT9313_0292\n",
       "NaN    PMIT9313_0294\n",
       "NaN    PMIT9303_0438\n",
       "Name: sseqid, dtype: object"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dout_df.sseqid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
